<!DOCTYPE html>
  <head>

 <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>Programming in R  | GEN242</title>
<link rel="stylesheet" type="text/css" href="../css/syntax.css">
<link rel="stylesheet" type="text/css" href="../css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="../css/bootstrap.min.css">-->
<link rel="stylesheet" type="text/css" href="../css/modern-business.css">
<link rel="stylesheet" type="text/css" href="../css/lavish-bootstrap.css">
<link rel="stylesheet" type="text/css" href="../css/customstyles.css">
<link rel="stylesheet" type="text/css" href="../css/theme-blue.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="../js/jquery.navgoco.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="../js/toc.js"></script>
<script src="../js/customscripts.js"></script>
<link rel="shortcut icon" href="../common_images/favicon.ico" type="image/x-icon">
<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->







 

<script>
  $(function () {
      $('[data-toggle="tooltip"]').tooltip()
  })
</script>



  </head>

<body>

   <!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <a class="fa fa-home fa-lg navbar-brand" href="home.html">&nbsp;<span class="projectTitle"> Data Analysis in Genome Biology</span></a>

        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- entries without drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                



















                
                
                
                
                <li><a href="https://piazza.com/class/ilbslqb3j2ovy" target="_blank">Piazza</a></li>
                
                
                
                
                
                <li><a href="https://github.com/tgirke/GEN242" target="_blank">Github Repo</a></li>
                
                
                
                


                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->

                <li class="dropdown">
                    
                    
                    
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://jekyllrb.com/" target="_blank">Jekyll</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                        
                        

                    </ul>
                </li>
                
                


                <!-- special insertion -->

               
                
                
                <!-- Send feedback function -->
<script>
function SendLinkByMail(href) {
var subject= "GEN242 feedback";
var body = "I have some feedback about the Programming in R page: ";
body += window.location.href;
body += "";
var uri = "mailto:?subject=";
uri += encodeURIComponent(subject);
uri += "&body=";
uri += encodeURIComponent(body);
window.location.href = uri;
}
</script>

<li><a href="javascript:(function()%7BSendLinkByMail()%3B%7D)()%3B" target="_blank"><i class="fa fa-envelope-o"></i> Feedback</a></li>


                <!--uncomment this block if you want simple search instead of algolia-->
                <li>
                     <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="../js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                        SimpleJekyllSearch.init({
                            searchInput: document.getElementById('search-input'),
                            resultsContainer: document.getElementById('results-container'),
                            dataSource: '../search.json',
                            searchResultTemplate: '<li><a href="{url}" title="Programming in R">{title}</a></li>',
                        noResultsText: 'No results found.',
                                limit: 10,
                                fuzzy: true,
                        })
                    </script>
                     <!--end search-->
                </li>

            
        </div>
        <!-- /.container -->
</nav>


    <!-- Page Content -->
    <div class="container">
        <div class="col-lg-12">&nbsp;</div>


<!-- Content Row -->
<div class="row">
    <!-- Sidebar Column -->
    <div class="col-md-3">

        <script>

            $(document).ready(function() {
                // Initialize navgoco with default options
                $("#mysidebar").navgoco({
                    caretHtml: '',
                    accordion: true,
                    openClass: 'active', // open
                    save: false, // leave false or nav highlighting doesn't work right
                    cookie: {
                        name: 'navgoco',
                        expires: false,
                        path: '/'
                    },
                    slide: {
                        duration: 400,
                        easing: 'swing'
                    }
                });

                $("#collapseAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', false);
                });

                $("#expandAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', true);
                });

            });

        </script>


        



















        <ul id="mysidebar" class="nav">

            <span class="siteTagline">GEN242</span>
            <span class="versionTagline">Spring 2016</span>

            
            
            
                
            
            <li><a href="#">Home</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/home.html">Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_syllabus.html">Syllabus</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Internal</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_internal.html">Internal Resources</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Schedule</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_schedule.html">Course Schedule</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Slides</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_01.html">Course Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_02.html">Molbiol for Bioinformatics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_03.html">Databases and Software</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_04.html">DNA Sequencing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_05.html">Introduction to R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_06.html">Sequence Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_07.html">Programming in R</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Homework</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_01.html">HW1 - Online Tools</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_02.html">HW2 - Biocluster/Linux</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_03.html">HW3 - Introduction to R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_04.html">HW4 - Pairwise Alignments</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">GitHub</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_tutorial_01.html">GitHub</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Biocluster</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_tutorial_02.html">Biocluster</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Introduction to R</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_03.html">3. Installation of R and Add-on Packages</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_04.html">4. Getting Around</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_06.html">6. Data Types</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_07.html">7. Data objects</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_08.html">8. Important Utilities</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_12.html">12. Graphics in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_13.html">13. Analysis Routine</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_14.html">14. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_15.html">15. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Programming in R</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_1.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_2.html">2. Control Structures</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_3.html">3. Loops</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_4.html">4. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_5.html">5. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Projects</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_01.html">Project 1</a></li>
                    

                    

                    
                    
                </ul>
                
                
                


                <!-- if you aren't using the accordion, uncomment this block:

                     <p class="external">
                         <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
                     </p>
                 -->

    </div>
    <!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
    <script>$("li.active").parents('li').toggleClass("active");</script>


            <!-- Content Column -->
            <div class="col-md-9">
                
                <div class="post-header">
   <h1 class="post-title-main">Programming in R</h1>
</div>

<div class="post-content">

   

<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>


<div id="toc"></div>


<!--
    
    <a target="_blank" href="https://github.com/tgirke/manuals/tree/master/mydoc/" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
    
-->
<!--
%% \VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Overview Vignette}
%% \VignetteDepends{methods}
%% \VignetteKeywords{compute cluster, pipeline, reports}
%% \VignettePackage{longevityTools}
-->

<!---
- Compile from command-line
echo "rmarkdown::render('Programming_in_R.Rmd', clean=F)" | R -slave; R CMD Stangle Programming_in_R.Rmd; Rscript ../md2jekyll.R Programming_in_R.knit.md 9

- Commit to github
git commit -am "some edits"; git push -u origin master

- To customize font size and other style features, add this line to output section in preamble:  
    css: style.css
-->

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

<h1 id="overview">Overview</h1>

<p>One of the main attractions of using the R
(<a href="http://cran.at.r-project.org">http://cran.at.r-project.org</a>) environment is
the ease with which users can write their own programs and custom functions.
The R programming syntax is extremely easy to learn, even for users with no
previous programming experience. Once the basic R programming control
structures are understood, users can use the R language as a powerful
environment to perform complex custom analyses of almost any type of data [@Gentleman2008-xo].</p>

<h2 id="why-programming-in-r">Why Programming in R?</h2>

<ul>
  <li>Powerful statistical environment and programming language</li>
  <li>Facilitates reproducible research</li>
  <li>Efficient data structures make programming very easy</li>
  <li>Ease of implementing custom functions</li>
  <li>Powerful graphics</li>
  <li>Access to fast growing number of analysis packages</li>
  <li>Most widely used language in bioinformatics</li>
  <li>Is standard for data mining and biostatistical analysis</li>
  <li>Technical advantages: free, open-source, available for all OSs</li>
</ul>

<h2 id="r-basics">R Basics</h2>

<p>The previous <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc/mydoc_Rbasics_01.html">Rbasics</a> provides a general introduction to the usage of the R environment and its basic command syntax.
More details can be found R &amp; BioConductor manual <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual">here</a>.</p>

<h2 id="code-editors-for-r">Code Editors for R</h2>

<p>Several excellent code editors are available that provide functionalities like R syntax highlighting, auto code indenting and utilities to send code/functions to the R console.</p>

<ul>
  <li><a href="https://www.rstudio.com/products/rstudio/features/">RStudio</a>: GUI-based IDE for R</li>
  <li><a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">Vim-R-Tmux</a>: R working environment based on vim and tmux</li>
  <li><a href="http://www.xemacs.org/Download/index.html">Emacs</a> (<a href="http://ess.r-project.org/">ESS add-on package</a>)</li>
  <li><a href="https://wiki.gnome.org/Apps/Gedit">gedit</a> and <a href="https://wiki.gnome.org/Apps/Gedit">Rgedit</a></li>
  <li><a href="https://rkward.kde.org/">RKWard</a></li>
  <li><a href="http://www.walware.de/goto/statet">Eclipse</a></li>
  <li><a href="http://jekyll.math.byuh.edu/other/howto/tinnr/install.shtml">Tinn-R</a></li>
  <li><a href="https://sourceforge.net/projects/npptor/">Notepad++ (NppToR)</a></li>
</ul>

<center> Programming in R using Vim or Emacs</center>
<center><img title="vim-r" src="images/vimR.png" /></center>

<center>Programming in R using RStudio</center>
<center><img title="R_Interfaces" src="images/rstudio.png" /></center>

<h2 id="finding-help">Finding Help</h2>

<p>Reference list on R programming (selection)</p>

<ul>
  <li><a href="http://adv-r.had.co.nz/">Advanced R</a>, by Hadley Wickham</li>
  <li><a href="http://master.bioconductor.org/help/publications/books/r-programming-for-bioinformatics/">R Programming for Bioinformatics</a>, by Robert Gentleman</li>
  <li><a href="http://www.stats.ox.ac.uk/pub/MASS3/Sprog/">S Programming</a>, by W. N. Venables and B. D. Ripley</li>
  <li><a href="http://www.amazon.com/Programming-Data-Language-Lecture-Economics/dp/0387985034">Programming with Data</a>, by John M. Chambers</li>
  <li><a href="http://www1.maths.lth.se/help/R/">R Help</a> &amp; <a href="http://www1.maths.lth.se/help/R/RCC/">R Coding Conventions</a>, Henrik Bengtsson, Lund University</li>
  <li><a href="http://zoonek2.free.fr/UNIX/48_R/02.html">Programming in R</a> (Vincent Zoonekynd)</li>
  <li><a href="http://www2.warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r">Peterâ€™s R Programming Pages</a>, University of Warwick</li>
  <li><a href="http://pj.freefaculty.org/R/statsRus.html">Rtips</a>, Paul Johnsson, University of Kansas</li>
  <li><a href="http://heather.cs.ucdavis.edu/~matloff/r.html">R for Programmers</a>, Norm Matloff, UC Davis</li>
  <li><a href="http://www.statistik.uni-dortmund.de/useR-2008/tutorials/useR2008introhighperfR.pdf">High-Performance R</a>, Dirk Eddelbuettel tutorial presented at useR-2008</li>
  <li><a href="http://www.stat.harvard.edu/ccr2005/index.html">C/C++ level programming for R</a>, Gopi Goswami</li>
</ul>

<h1 id="control-structures">Control Structures</h1>

<h2 id="important-operators">Important Operators</h2>

<h3 id="comparison-operators">Comparison operators</h3>

<ul>
  <li><code>==</code> (equal)</li>
  <li><code>!=</code> (not equal)</li>
  <li><code>&gt;</code> (greater than)</li>
  <li><code>&gt;=</code> (greater than or equal)</li>
  <li><code>&lt;</code> (less than)</li>
  <li><code>&lt;=</code> (less than or equal)</li>
</ul>

<h3 id="logical-operators">Logical operators</h3>

<ul>
  <li><code>&amp;</code> (and)</li>
  <li><code>|</code> (or)</li>
  <li><code>!</code> (not)</li>
</ul>

<h2 id="conditional-executions-if-statements">Conditional Executions: <code>if</code> Statements</h2>

<p>An <code>if</code> statement operates on length-one logical vectors.</p>

<p><strong>Syntax</strong></p>

<p><code>r
if(TRUE) { 
	statements_1 
} else { 
	statements_2 
}
</code></p>

<p><strong>Example</strong></p>

<p><code>r
if(1==0) { 
	print(1) 
} else { 
	print(2) 
}
</code></p>

<p><code>
## [1] 2
</code></p>

<h2 id="conditional-executions-ifelse-statements">Conditional Executions: <code>ifelse</code> Statements`</h2>

<p>The <code>ifelse</code> statement operates on vectors.</p>

<p><strong>Syntax</strong></p>

<p><code>r
ifelse(test, true_value, false_value)
</code>
<strong>Example</strong></p>

<p><code>r
x &lt;- 1:10 
ifelse(x&lt;5, x, 0)
</code></p>

<p><code>
##  [1] 1 2 3 4 0 0 0 0 0 0
</code></p>

<h1 id="loops">Loops</h1>

<h2 id="for-loop"><code>for</code> loop</h2>

<p><code>for</code> loops iterate over elements of a looping vector.</p>

<p><strong>Syntax</strong></p>

<p><code>r
for(variable in sequence) { 
	statements 
}
</code>
<strong>Example</strong></p>

<p><code>r
mydf &lt;- iris
myve &lt;- NULL
for(i in seq(along=mydf[,1])) {
	myve &lt;- c(myve, mean(as.numeric(mydf[i,1:3])))
}
myve[1:8]
</code></p>

<p><code>
## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></p>

<p><strong>Note:</strong> Inject into objecs is much faster than append approach with <code>c</code>, <code>cbind</code>, etc.</p>

<p><strong>Example</strong></p>

<p><code>r
myve &lt;- numeric(length(mydf[,1]))
for(i in seq(along=myve)) {
	myve[i] &lt;- mean(as.numeric(mydf[i,1:3]))
}
myve[1:8]
</code></p>

<p><code>
## [1] 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></p>

<h3 id="conditional-stop-of-loops">Conditional Stop of Loops</h3>

<p>The <code>stop</code> function can be used to break out of a loop (or a function) when a condition becomes <code>TRUE</code>. In addition, an error message will be printed.</p>

<p><strong>Example</strong></p>

<p><code>r
x &lt;- 1:10
z &lt;- NULL
for(i in seq(along=x)) { 
	if(x[i] &lt; 5) { 
		z &lt;- c(z, x[i]-1)  
	} else { 
		stop("values need to be &lt; 5") 
	}
}
</code></p>

<h2 id="while-loop"><code>while</code> loop</h2>

<p>Iterates as long as a condition is true.</p>

<p><strong>Syntax</strong></p>

<p><code>r
while(condition) {
	statements
}
</code></p>

<p><strong>Example</strong></p>

<p><code>r
z &lt;- 0
while(z&lt;5) { 
	z &lt;- z + 2
	print(z)  
}
</code></p>

<p><code>
## [1] 2
## [1] 4
## [1] 6
</code></p>

<h2 id="the-apply-function-family">The <code>apply</code> Function Family</h2>

<h3 id="apply"><code>apply</code></h3>

<p><strong>Syntax</strong></p>

<p><code>r
apply(X, MARGIN, FUN, ARGs)
</code></p>

<p><strong>Arguments</strong></p>

<ul>
  <li><code>X</code>: <code>array</code>, <code>matrix</code> or <code>data.frame</code></li>
  <li><code>MARGIN</code>: <code>1</code> for rows, <code>2</code> for columns</li>
  <li><code>FUN</code>: one or more functions</li>
  <li><code>ARGs</code>: possible arguments for functions</li>
</ul>

<p><strong>Example</strong></p>

<p><code>r
apply(iris[1:8,1:3], 1, mean)
</code></p>

<p><code>
##        1        2        3        4        5        6        7        8 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667 3.133333 3.300000
</code></p>

<h3 id="tapply"><code>tapply</code></h3>

<p>Applies a function to vector components that are defined by a factor.</p>

<p><strong>Syntax</strong></p>

<p><code>r
tapply(vector, factor, FUN)
</code></p>

<p><strong>Example</strong></p>

<p><code>r
iris[1:2,]
</code></p>

<p><code>
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
</code></p>

<p><code>r
tapply(iris$Sepal.Length, iris$Species, mean)
</code></p>

<p><code>
##     setosa versicolor  virginica 
##      5.006      5.936      6.588
</code></p>

<h3 id="sapply-and-lapply"><code>sapply</code> and <code>lapply</code></h3>

<p>Both apply a function to vector or list objects. The <code>lapply</code> function always returns a list object, while <code>sapply</code> returns <code>vector</code> or <code>matrix</code> objects when it is possible.</p>

<p><strong>Examples</strong></p>

<p><code>r
x &lt;- list(a = 1:10, beta = exp(-3:3), logic = c(TRUE,FALSE,FALSE,TRUE))
lapply(x, mean)
</code></p>

<p><code>
## $a
## [1] 5.5
## 
## $beta
## [1] 4.535125
## 
## $logic
## [1] 0.5
</code></p>

<p><code>r
sapply(x, mean)
</code></p>

<p><code>
##        a     beta    logic 
## 5.500000 4.535125 0.500000
</code></p>

<p>Often used in combination with a function definition:</p>

<p><code>r
lapply(names(x), function(x) mean(x))
sapply(names(x), function(x) mean(x))
</code></p>

<h1 id="session-info">Session Info</h1>

<p><code>r
sessionInfo()
</code></p>

<p><code>
## R version 3.2.4 Revised (2016-03-16 r70336)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.4 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  utils     datasets  grDevices methods   base     
## 
## other attached packages:
## [1] ggplot2_2.0.0   limma_3.26.3    BiocStyle_1.8.0
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.3      codetools_0.2-14 digest_0.6.9     plyr_1.8.3       grid_3.2.4      
##  [6] gtable_0.1.2     formatR_1.2.1    magrittr_1.5     evaluate_0.8     scales_0.3.0    
## [11] stringi_1.0-1    rmarkdown_0.9.2  tools_3.2.4      stringr_1.0.0    munsell_0.4.2   
## [16] yaml_2.1.13      colorspace_1.2-6 htmltools_0.3    knitr_1.12
</code></p>

<h1 id="references">References</h1>



<div class="tags">
    
</div>



</div>



<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2016 UC Riverside. All rights reserved. <br />
 Site last generated: Apr 13, 2016 <br />
<p><img src="../common_images/company_logo.png/></p>
                </div>
            </div>
</footer>

            </div><!-- /.row -->

    </div>    <!-- /.container -->

</body>


</html>

