<!DOCTYPE html>
  <head>

 <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>NGS Analysis Basics  | GEN242</title>
<link rel="stylesheet" type="text/css" href="../css/syntax.css">
<link rel="stylesheet" type="text/css" href="../css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="../css/bootstrap.min.css">-->
<link rel="stylesheet" type="text/css" href="../css/modern-business.css">
<link rel="stylesheet" type="text/css" href="../css/lavish-bootstrap.css">
<link rel="stylesheet" type="text/css" href="../css/customstyles.css">
<link rel="stylesheet" type="text/css" href="../css/theme-blue.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="../js/jquery.navgoco.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="../js/toc.js"></script>
<script src="../js/customscripts.js"></script>
<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<link rel="shortcut icon" href="../common_images/favicon.ico" type="image/x-icon">
<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->







 


<script>
  $(function () {
      $('[data-toggle="tooltip"]').tooltip()
  })
</script>



  </head>

<body>

   <!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <a class="fa fa-home fa-lg navbar-brand" href="home.html">&nbsp;<span class="projectTitle"> Data Analysis in Genome Biology</span></a>

        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- entries without drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                









                
                
                
                
                <li><a href="https://piazza.com/class/ilbslqb3j2ovy" target="_blank">Piazza</a></li>
                
                
                
                
                
                <li><a href="https://github.com/tgirke/GEN242" target="_blank">Github Repo</a></li>
                
                
                
                


                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->

                <li class="dropdown">
                    
                    
                    
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://jekyllrb.com/" target="_blank">Jekyll</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                        
                        

                    </ul>
                </li>
                
                


                <!-- special insertion -->

               
                <!-- Send feedback function -->
<script>
function SendLinkByMail(href) {
var subject= "GEN242 feedback";
var body = "I have some feedback about the NGS Analysis Basics page: ";
body += window.location.href;
body += "";
var uri = "mailto:?subject=";
uri += encodeURIComponent(subject);
uri += "&body=";
uri += encodeURIComponent(body);
window.location.href = uri;
}
</script>

<li><a href="javascript:(function()%7BSendLinkByMail()%3B%7D)()%3B" target="_blank"><i class="fa fa-envelope-o"></i> Feedback</a></li>


                <!--uncomment this block if you want simple search instead of algolia-->
                <li>
                     <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="../js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                        SimpleJekyllSearch.init({
                            searchInput: document.getElementById('search-input'),
                            resultsContainer: document.getElementById('results-container'),
                            dataSource: '../search.json',
                            searchResultTemplate: '<li><a href="{url}" title="NGS Analysis Basics">{title}</a></li>',
                        noResultsText: 'No results found.',
                                limit: 10,
                                fuzzy: true,
                        })
                    </script>
                     <!--end search-->
                </li>

            
        </div>
        <!-- /.container -->
</nav>


    <!-- Page Content -->
    <div class="container">
        <div class="col-lg-12">&nbsp;</div>


<!-- Content Row -->
<div class="row">
    <!-- Sidebar Column -->
    <div class="col-md-3">

        <script>

            $(document).ready(function() {
                // Initialize navgoco with default options
                $("#mysidebar").navgoco({
                    caretHtml: '',
                    accordion: true,
                    openClass: 'active', // open
                    save: false, // leave false or nav highlighting doesn't work right
                    cookie: {
                        name: 'navgoco',
                        expires: false,
                        path: '/'
                    },
                    slide: {
                        duration: 400,
                        easing: 'swing'
                    }
                });

                $("#collapseAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', false);
                });

                $("#expandAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', true);
                });

            });

        </script>


        









        <ul id="mysidebar" class="nav">

            <span class="siteTagline">GEN242</span>
            <span class="versionTagline">Spring 2016</span>

            
            
            
                
            
            <li><a href="#">Home</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/home.html">Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_syllabus.html">Syllabus</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Internal</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_internal.html">Internal Resources</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Schedule</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_schedule.html">Course Schedule</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Lecture Slides</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_01.html">Course Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_02.html">Molbiol for Bioinformatics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_03.html">Databases and Software</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_04.html">DNA Sequencing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_05.html">Introduction to R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_06.html">Sequence Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_07.html">Programming in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_08.html">Multiple Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_09.html">Short Read Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_10.html">NGS Analysis Basics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_11.html">Gene Expression Analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_12.html">NGS Workflows</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_13.html">ChIP-Seq Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_14.html">VAR-Seq Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_15.html">Sequence Assembly</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_16.html">Gene Annotation and Ontologies</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_17.html">Cluster Analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_18.html">Profile HMMs</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_19.html">Introduction to Phylogenetics</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Homework</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_01.html">HW1 - Online Tools</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_02.html">HW2 - Biocluster/Linux</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_03.html">HW3 - Introduction to R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_04.html">HW4 - Pairwise Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_09.html">HW5 - Programming in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_08.html">HW6 - Sequence Analysis</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">GitHub</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_tutorial_01.html">GitHub</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Biocluster & Linux</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_tutorial_02.html">Basic Tutorial</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Introduction to R</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_03.html">3. Installation of R and Add-on Packages</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_04.html">4. Getting Around</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_06.html">6. Data Types</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_07.html">7. Data objects</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_08.html">8. Important Utilities</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_12.html">12. SQLite Databases</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_13.html">13. Graphics in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_14.html">14. Analysis Routine</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_15.html">15. R Markdown</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_16.html">16. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_17.html">17. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Programming in R</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_02.html">2. Control Structures</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_03.html">3. Loops</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_04.html">4. Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_05.html">5. Useful Utilities</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_06.html">6. Running R Scripts</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_07.html">7. Building R Packages</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_08.html">8. Programming Exercises</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_09.html">9. Homework 5</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_10.html">10. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_11.html">11. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">NGS Analysis Basics</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_02.html">2. Package Requirements</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_03.html">3. Strings in R Base</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_04.html">4. Sequences in Bioconductor</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_05.html">5. NGS Sequences</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_06.html">6. Range Operations</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_07.html">7. Transcript Ranges</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_08.html">8. Homework 6</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_09.html">9. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_10.html">10. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">NGS Workflow Overview</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_1.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_2.html">2. Getting Started</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_5.html">5. Version information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_6.html">6. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">RNA-Seq Workflow</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_02.html">2. Sample definitions and environment settings</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_05.html">5. Read quantification per annotation range</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_06.html">6. Analysis of differentially expressed genes with edgeR</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis of DEGs</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_11.html">11. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">ChIP-Seq Workflow</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_02.html">2. Load workflow environment</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_14.html">14. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">VAR-Seq Workflow</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_02.html">2. Load workflow environment</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_14.html">14. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Graphics and Visualization</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_1.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_2.html">2. Base Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_3.html">3. Grid Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_4.html">4. lattice Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_5.html">5. ggplot2 Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_6.html">6. Specialty Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_7.html">7. Genome Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_8.html">8. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Cluster Analysis</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_1.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_2.html">2. Data Preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_3.html">3. Clustering Algorithms</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_4.html">4. Clustering Exercises</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_5.html">5. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_6.html">6. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Paper Presentations</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_paper_presentations.html">Paper Presentations</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Course Projects</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_00.html">Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_01.html">ChIP-Seq1 - Motif Prediction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_02.html">ChIP-Seq2 - Peak Callers</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_03.html">RNA-Seq1 - RNA-Seq Aligners</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_04.html">RNA-Seq2 - DEG Methods</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_05.html">VAR-Seq1 - Variant Callers</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_06.html">VAR-Seq2 - Functional Prediction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_07.html">Project Data</a></li>
                    

                    

                    
                    
                </ul>
                
                
                


                <!-- if you aren't using the accordion, uncomment this block:

                     <p class="external">
                         <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
                     </p>
                 -->

    </div>
    <!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
    <script>$("li.active").parents('li').toggleClass("active");</script>


            <!-- Content Column -->
            <div class="col-md-9">
                
                <div class="post-header">
   <h1 class="post-title-main">NGS Analysis Basics</h1>
</div>

<div class="post-content">

   

<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>


<div id="toc"></div>


<!--
    
    <a target="_blank" href="https://github.com/tgirke/manuals/tree/master/mydoc/" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
    
-->
<!--
%% \VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Overview Vignette}
%% \VignetteDepends{methods}
%% \VignetteKeywords{compute cluster, pipeline, reports}
%% \VignettePackage{longevityTools}
-->

<!---
- Compile from command-line
echo "rmarkdown::render('Rsequences.Rmd', clean=F)" | R -slave; R CMD Stangle Rsequences.Rmd; Rscript ../md2jekyll.R Rsequences.knit.md 10

- Commit to github
git commit -am "some edits"; git push -u origin master

- To customize font size and other style features, add this line to output section in preamble:  
    css: style.css
-->

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

<h1 id="overview">Overview</h1>

<h2 id="sequence-analysis-in-r-and-bioconductor">Sequence Analysis in R and Bioconductor</h2>

<p><strong>R Base</strong></p>

<ul>
  <li>Some basic string handling utilities. Wide spectrum of numeric data analysis tools.</li>
</ul>

<p><strong>Bioconductor</strong></p>

<p>Bioconductor packages provide much more sophisticated string handling utilities for sequence analysis [@Lawrence2013-kt, @Huber2015-ag].</p>

<ul>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a>: general sequence analysis environment</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/ShortRead.html">ShortRead</a>: pipeline for short read data</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a>: low-level infrastructure for range data</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomicRanges</a>: high-level infrastructure for range data</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html">GenomicFeatures</a>: managing transcript centric annotations</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/GenomicAlignments.html">GenomicAlignments</a>: handling short genomic alignments</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/Rsamtools.html">Rsamtools</a>: interface to  <code class="highlighter-rouge">samtools</code>, <code class="highlighter-rouge">bcftools</code> and <code class="highlighter-rouge">tabix</code></li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/BSgenome.html">BSgenome</a>: genome annotation data</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a>: interface to BioMart annotations</li>
  <li><a href="http://bioconductor.org/packages/release/bioc/html/rtracklayer.html">rtracklayer</a>: Annotation imports, interface to online genome browsers</li>
</ul>

<h1 id="package-requirements">Package Requirements</h1>

<p>Several Bioconductor packages are required for this tutorial. To install them, execute
the following lines in the R console. Please also make sure that you have a recent R version
installed on your system. R versions <code class="highlighter-rouge">3.2.x</code> or higher are recommended.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">source</span><span class="p">(</span><span class="s2">"https://bioconductor.org/biocLite.R"</span><span class="p">)</span>
<span class="n">biocLite</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"Biostrings"</span><span class="p">,</span> <span class="s2">"GenomicRanges"</span><span class="p">,</span> <span class="s2">"GenomicRanges"</span><span class="p">,</span> <span class="s2">"rtracklayer"</span><span class="p">,</span> <span class="s2">"systemPipeR"</span><span class="p">,</span> <span class="s2">"seqLogo"</span><span class="p">,</span> <span class="s2">"ShortRead"</span><span class="p">))</span>
</code></pre>
</div>

<h1 id="strings-in-r-base">Strings in R Base</h1>

<h2 id="basic-string-matching-and-parsing">Basic String Matching and Parsing</h2>

<h3 id="string-matching">String matching</h3>

<p>Generate sample sequence data set</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">myseq</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">"ATGCAGACATAGTG"</span><span class="p">,</span> <span class="s2">"ATGAACATAGATCC"</span><span class="p">,</span> <span class="s2">"GTACAGATCAC"</span><span class="p">)</span>
</code></pre>
</div>

<p>String searching with regular expression support</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">myseq</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"ATG"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">)]</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "ATGCAGACATAGTG" "ATGAACATAGATCC"
</code></pre>
</div>

<p>Searches <code class="highlighter-rouge">myseq</code> for first match of pattern AT</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">pos1</span> <span class="o">&lt;-</span> <span class="n">regexpr</span><span class="p">(</span><span class="s2">"AT"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">)</span> 
<span class="n">as.numeric</span><span class="p">(</span><span class="n">pos1</span><span class="p">);</span> <span class="n">attributes</span><span class="p">(</span><span class="n">pos1</span><span class="p">)</span><span class="o">$</span><span class="n">match.length</span> <span class="c1"># Returns position information of matches
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 1 7
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 2 2 2
</code></pre>
</div>

<p>Searches <code class="highlighter-rouge">myseq</code> for all matches of pattern AT</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">pos2</span> <span class="o">&lt;-</span> <span class="n">gregexpr</span><span class="p">(</span><span class="s2">"AT"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">)</span> 
<span class="n">as.numeric</span><span class="p">(</span><span class="n">pos2</span><span class="p">[[</span><span class="m">1</span><span class="p">]]);</span> <span class="n">attributes</span><span class="p">(</span><span class="n">pos2</span><span class="p">[[</span><span class="m">1</span><span class="p">]])</span><span class="o">$</span><span class="n">match.length</span> <span class="c1"># Returns positions of matches in first sequence
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 9
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 2 2
</code></pre>
</div>

<p>String substitution with regular expression support</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gsub</span><span class="p">(</span><span class="s2">"^ATG"</span><span class="p">,</span> <span class="s2">"atg"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">)</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "atgCAGACATAGTG" "atgAACATAGATCC" "GTACAGATCAC"
</code></pre>
</div>

<h3 id="positional-parsing">Positional parsing</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">nchar</span><span class="p">(</span><span class="n">myseq</span><span class="p">)</span> <span class="c1"># Computes length of strings
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 14 14 11
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">substring</span><span class="p">(</span><span class="n">myseq</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="n">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">3</span><span class="p">),</span> <span class="n">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">5</span><span class="p">))</span> <span class="c1"># Positional parsing of several fragments from one string
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "AT"  "GCA"
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">substring</span><span class="p">(</span><span class="n">myseq</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">4</span><span class="p">,</span><span class="m">7</span><span class="p">),</span> <span class="n">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">6</span><span class="p">,</span><span class="m">10</span><span class="p">))</span> <span class="c1"># Positional parsing of many strings
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "AT"   "AAC"  "ATCA"
</code></pre>
</div>

<h2 id="random-sequence-generation">Random Sequence Generation</h2>

<h3 id="random-dna-sequences-of-any-length">Random DNA sequences of any length</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">rand</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">,</span> <span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">paste</span><span class="p">(</span><span class="n">sample</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"A"</span><span class="p">,</span><span class="s2">"T"</span><span class="p">,</span><span class="s2">"G"</span><span class="p">,</span><span class="s2">"C"</span><span class="p">),</span> <span class="n">sample</span><span class="p">(</span><span class="m">10</span><span class="o">:</span><span class="m">20</span><span class="p">),</span> <span class="n">replace</span><span class="o">=</span><span class="n">T</span><span class="p">),</span> <span class="n">collapse</span><span class="o">=</span><span class="s2">""</span><span class="p">))</span>
<span class="n">rand</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "TTGGTGCCCGCACACGT" "TACTCACGAC"        "AGCTCAACCTAGTTTCG"
</code></pre>
</div>

<h3 id="count-identical-sequences">Count identical sequences</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">table</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">rand</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span> <span class="n">rand</span><span class="p">[</span><span class="m">1</span><span class="p">]))</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## 
##    AGCTCAACCTAGTTTCG CGCTGCTTGGCCCAGCTGAC           TACTCACGAC    TTGGTGCCCGCACACGT 
##                    1                    1                    1                    2
</code></pre>
</div>

<h3 id="extract-reads-from-reference">Extract reads from reference</h3>

<p>Note: this requires <code class="highlighter-rouge">Biostrings</code> package.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">Biostrings</span><span class="p">)</span>
<span class="n">ref</span> <span class="o">&lt;-</span> <span class="n">DNAString</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">sample</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"A"</span><span class="p">,</span><span class="s2">"T"</span><span class="p">,</span><span class="s2">"G"</span><span class="p">,</span><span class="s2">"C"</span><span class="p">),</span> <span class="m">100000</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="n">T</span><span class="p">),</span> <span class="n">collapse</span><span class="o">=</span><span class="s2">""</span><span class="p">))</span>
<span class="n">randstart</span> <span class="o">&lt;-</span> <span class="n">sample</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">ref</span><span class="p">)</span><span class="m">-15</span><span class="p">),</span> <span class="m">1000</span><span class="p">)</span>
<span class="n">randreads</span> <span class="o">&lt;-</span> <span class="n">Views</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">randstart</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="m">15</span><span class="p">)</span>
<span class="n">rand_set</span> <span class="o">&lt;-</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">randreads</span><span class="p">)</span>
<span class="n">unlist</span><span class="p">(</span><span class="n">rand_set</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   15000-letter "DNAString" instance
## seq: TTTTTTAAAGATCGTTCTACACCGGATGCCTTGCTGTGCTGATCTT...CCGTATTCCAACTATACTCCTCCGACCGTTAGATCGGTAGGTGCGG
</code></pre>
</div>

<h1 id="sequences-in-bioconductor">Sequences in Bioconductor</h1>

<h2 id="important-data-objects-of-biostrings">Important Data Objects of Biostrings</h2>

<h3 id="xstring-for-single-sequence"><code class="highlighter-rouge">XString</code> for single sequence</h3>

<ul>
  <li><code class="highlighter-rouge">DNAString</code>: for DNA</li>
  <li><code class="highlighter-rouge">RNAString</code>: for RNA</li>
  <li><code class="highlighter-rouge">AAString</code>: for amino acid</li>
  <li><code class="highlighter-rouge">BString</code>: for any string</li>
</ul>

<h3 id="xstringset-for-many-sequences"><code class="highlighter-rouge">XStringSet</code> for many sequences</h3>

<ul>
  <li><code class="highlighter-rouge">DNAStringSet</code>`: for DNA</li>
  <li><code class="highlighter-rouge">RNAStringSet</code>: for RNA</li>
  <li><code class="highlighter-rouge">AAStringSet</code>: for amino acid</li>
  <li><code class="highlighter-rouge">BStringSet</code>: for any string</li>
</ul>

<h3 id="qualityscalexstringset-for-sequences-with-quality-data"><code class="highlighter-rouge">QualityScaleXStringSet</code> for sequences with quality data</h3>

<ul>
  <li><code class="highlighter-rouge">QualityScaledDNAStringSet</code>: for DNA</li>
  <li><code class="highlighter-rouge">QualityScaledRNAStringSet</code>: for RNA</li>
  <li><code class="highlighter-rouge">QualityScaledAAStringSet</code>: for amino acid</li>
  <li><code class="highlighter-rouge">QualityScaledBStringSet</code>: for any string</li>
</ul>

<h2 id="sequence-import-and-export">Sequence Import and Export</h2>

<p>Download the following sequences to your current working directory and then import them into R: 
<a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn">ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn</a></p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">dir.create</span><span class="p">(</span><span class="s2">"data"</span><span class="p">,</span> <span class="n">showWarnings</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
<span class="c1"># system("wget ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn")
</span><span class="n">download.file</span><span class="p">(</span><span class="s2">"ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.ffn"</span><span class="p">,</span> <span class="s2">"data/AE004437.ffn"</span><span class="p">)</span>
</code></pre>
</div>

<p>Import FASTA file with <code class="highlighter-rouge">readDNAStringSet</code></p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">myseq</span> <span class="o">&lt;-</span> <span class="n">readDNAStringSet</span><span class="p">(</span><span class="s2">"data/AE004437.ffn"</span><span class="p">)</span>
<span class="n">myseq</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 3
##     width seq                                                                   names               
## [1]  1206 ATGACTCGGCGGTCTCGTGTCGGTGCCGGCCTC...GTCGTCGTTGTTCGACGCTGGCGGAACCCATGA gi|12057215|gb|AE...
## [2]   666 ATGAGCATCATCGAACTCGAAGGCGTGGTCAAA...GTCAACCTCGTCGATGGGGTGTTACACACGTGA gi|12057215|gb|AE...
## [3]  1110 ATGGCGTGGCGGAACCTCGGGCGGAACCGCGTG...AACGATCCGCCCGTCGAGGCGCTCGGCGAATGA gi|12057215|gb|AE...
</code></pre>
</div>

<p>Subset sequences with regular expression on sequence name field</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sub</span> <span class="o">&lt;-</span> <span class="n">myseq</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"99.*"</span><span class="p">,</span> <span class="n">names</span><span class="p">(</span><span class="n">myseq</span><span class="p">))]</span>
<span class="n">length</span><span class="p">(</span><span class="n">sub</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 170
</code></pre>
</div>

<p>Export subsetted sequences to FASTA file</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">writeXStringSet</span><span class="p">(</span><span class="n">sub</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">"./data/AE004437sub.ffn"</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="m">80</span><span class="p">)</span>
</code></pre>
</div>

<p>Now inspect exported sequence file <code class="highlighter-rouge">AE004437sub.ffn</code> in a text editor</p>

<h2 id="working-with-xstring-containers">Working with <code class="highlighter-rouge">XString</code> Containers</h2>

<p>The <code class="highlighter-rouge">XString</code> stores the different types of biosequences in dedicated containers</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">Biostrings</span><span class="p">)</span>
<span class="n">d</span> <span class="o">&lt;-</span> <span class="n">DNAString</span><span class="p">(</span><span class="s2">"GCATAT-TAC"</span><span class="p">)</span>
<span class="n">d</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   10-letter "DNAString" instance
## seq: GCATAT-TAC
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">d</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   4-letter "DNAString" instance
## seq: GCAT
</code></pre>
</div>

<p>RNA sequences</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">r</span> <span class="o">&lt;-</span> <span class="n">RNAString</span><span class="p">(</span><span class="s2">"GCAUAU-UAC"</span><span class="p">)</span> 
<span class="n">r</span> <span class="o">&lt;-</span> <span class="n">RNAString</span><span class="p">(</span><span class="n">d</span><span class="p">)</span> <span class="c1"># Converts d to RNAString object
</span><span class="n">r</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   10-letter "RNAString" instance
## seq: GCAUAU-UAC
</code></pre>
</div>
<p>Protein sequences</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">p</span> <span class="o">&lt;-</span> <span class="n">AAString</span><span class="p">(</span><span class="s2">"HCWYHH"</span><span class="p">)</span>
<span class="n">p</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   6-letter "AAString" instance
## seq: HCWYHH
</code></pre>
</div>

<p>Any type of character strings</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">b</span> <span class="o">&lt;-</span> <span class="n">BString</span><span class="p">(</span><span class="s2">"I store any set of characters. Other XString objects store only the IUPAC characters."</span><span class="p">)</span>
<span class="n">b</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   85-letter "BString" instance
## seq: I store any set of characters. Other XString objects store only the IUPAC characters.
</code></pre>
</div>

<h2 id="working-with-xstringset-containers">Working with <code class="highlighter-rouge">XStringSet</code> Containers</h2>

<p><code class="highlighter-rouge">XStringSet</code> containers allow to store many biosequences in one object</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">dset</span> <span class="o">&lt;-</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"GCATATTAC"</span><span class="p">,</span> <span class="s2">"AATCGATCC"</span><span class="p">,</span> <span class="s2">"GCATATTAC"</span><span class="p">))</span> 
<span class="n">names</span><span class="p">(</span><span class="n">dset</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">"seq1"</span><span class="p">,</span> <span class="s2">"seq2"</span><span class="p">,</span> <span class="s2">"seq3"</span><span class="p">)</span> <span class="c1"># Assigns names
</span><span class="n">dset</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq                                                                   names               
## [1]     9 GCATATTAC                                                             seq1
## [2]     9 AATCGATCC                                                             seq2
</code></pre>
</div>

<p>Important utilities for <code class="highlighter-rouge">XStringSet</code> containers</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">width</span><span class="p">(</span><span class="n">dset</span><span class="p">)</span> <span class="c1"># Returns the length of each sequences
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 9 9 9
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">d</span> <span class="o">&lt;-</span> <span class="n">dset</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span> <span class="c1"># The [[ subsetting operator returns a single entry as XString object
</span><span class="n">dset2</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">dset</span><span class="p">,</span> <span class="n">dset</span><span class="p">)</span> <span class="c1"># Appends/concatenates two XStringSet objects
</span><span class="n">dsetchar</span> <span class="o">&lt;-</span> <span class="n">as.character</span><span class="p">(</span><span class="n">dset</span><span class="p">)</span> <span class="c1"># Converts XStringSet to named vector 
</span><span class="n">dsetone</span> <span class="o">&lt;-</span> <span class="n">unlist</span><span class="p">(</span><span class="n">dset</span><span class="p">)</span> <span class="c1"># Collapses many sequences to a single one stored in a DNAString container
</span></code></pre>
</div>

<p>Sequence subsetting by positions:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">DNAStringSet</span><span class="p">(</span><span class="n">dset</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">),</span> <span class="n">end</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="m">4</span><span class="p">,</span><span class="m">8</span><span class="p">,</span><span class="m">5</span><span class="p">))</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 3
##     width seq                                                                   names               
## [1]     4 GCAT                                                                  seq1
## [2]     7 ATCGATC                                                               seq2
## [3]     3 ATA                                                                   seq3
</code></pre>
</div>

<h2 id="multiple-alignment-class">Multiple Alignment Class</h2>

<p>The <code class="highlighter-rouge">XMultipleAlignment</code> class stores the different types of multiple sequence alignments:</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">origMAlign</span> <span class="o">&lt;-</span> <span class="n">readDNAMultipleAlignment</span><span class="p">(</span><span class="n">filepath</span> <span class="o">=</span> <span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span>
              <span class="s2">"msx2_mRNA.aln"</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s2">"Biostrings"</span><span class="p">),</span> <span class="n">format</span> <span class="o">=</span> <span class="s2">"clustal"</span><span class="p">)</span>
<span class="n">origMAlign</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## DNAMultipleAlignment with 8 rows and 2343 columns
##      aln                                                                        names               
## [1] -----TCCCGTCTCCGCAGCAAAAAAGTTTGAGTCG...TTGTCCAAACTCACAATTAAAAAAAAAAAAAAAAA gi|84452153|ref|N...
## [2] ------------------------------------...----------------------------------- gi|208431713|ref|...
## [3] ------------------------------------...----------------------------------- gi|118601823|ref|...
## [4] ----------------------AAAAGTTGGAGTCT...----------------------------------- gi|114326503|ref|...
## [5] ------------------------------------...----------------------------------- gi|119220589|ref|...
## [6] ------------------------------------...----------------------------------- gi|148540149|ref|...
## [7] --------------CGGCTCCGCAGCGCCTCACTCG...----------------------------------- gi|45383056|ref|N...
## [8] GGGGGAGACTTCAGAAGTTGTTGTCCTCTCCGCTGA...----------------------------------- gi|213515133|ref|...
</code></pre>
</div>

<h2 id="basic-sequence-manipulations">Basic Sequence Manipulations</h2>

<h3 id="reverse-and-complement">Reverse and Complement</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">randset</span> <span class="o">&lt;-</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">rand</span><span class="p">)</span>
<span class="n">complement</span><span class="p">(</span><span class="n">randset</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">])</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]    17 AACCACGGGCGTGTGCA
## [2]    10 ATGAGTGCTG
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">reverse</span><span class="p">(</span><span class="n">randset</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">])</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]    17 TGCACACGCCCGTGGTT
## [2]    10 CAGCACTCAT
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">reverseComplement</span><span class="p">(</span><span class="n">randset</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">])</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]    17 ACGTGTGCGGGCACCAA
## [2]    10 GTCGTGAGTA
</code></pre>
</div>

<h2 id="translate-dna-into-protein">Translate DNA into Protein</h2>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">translate</span><span class="p">(</span><span class="n">randset</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">])</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning in .Call2("DNAStringSet_translate", x, skip_code, dna_codes[codon_alphabet], : in 'x[[1]]':
## last 2 bases were ignored
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning in .Call2("DNAStringSet_translate", x, skip_code, dna_codes[codon_alphabet], : in 'x[[2]]':
## last base was ignored
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A AAStringSet instance of length 2
##     width seq
## [1]     5 LVPAH
## [2]     3 YSR
</code></pre>
</div>

<h2 id="pattern-matching">Pattern Matching</h2>

<h3 id="pattern-matching-with-mismatches">Pattern matching with mismatches</h3>

<p>Find pattern matches in reference</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">myseq1</span> <span class="o">&lt;-</span> <span class="n">readDNAStringSet</span><span class="p">(</span><span class="s2">"./data/AE004437.ffn"</span><span class="p">)</span> 
<span class="n">mypos</span> <span class="o">&lt;-</span> <span class="n">matchPattern</span><span class="p">(</span><span class="s2">"ATGGTG"</span><span class="p">,</span> <span class="n">myseq1</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span> <span class="n">max.mismatch</span><span class="o">=</span><span class="m">1</span><span class="p">)</span> 
</code></pre>
</div>

<p>Count only the corresponding matches</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">countPattern</span><span class="p">(</span><span class="s2">"ATGGCT"</span><span class="p">,</span> <span class="n">myseq1</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span> <span class="n">max.mismatch</span><span class="o">=</span><span class="m">1</span><span class="p">)</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 3
</code></pre>
</div>

<p>Count matches in many sequences</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">vcountPattern</span><span class="p">(</span><span class="s2">"ATGGCT"</span><span class="p">,</span> <span class="n">myseq1</span><span class="p">,</span> <span class="n">max.mismatch</span><span class="o">=</span><span class="m">1</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] 3 0 5 4 1 2 2 1 4 3 0 0 1 2 0 1 4 0 0 1
</code></pre>
</div>

<p>Results shown in DNAStringSet object</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">tmp</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">DNAStringSet</span><span class="p">(</span><span class="s2">"ATGGTG"</span><span class="p">),</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">mypos</span><span class="p">))</span> 
</code></pre>
</div>

<p>Return a consensus  matrix for query and hits</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">consensusMatrix</span><span class="p">(</span><span class="n">tmp</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   [,1] [,2] [,3] [,4] [,5] [,6]
## A    3    0    0    0    0    0
## C    1    1    0    0    0    0
## G    0    0    4    4    1    4
## T    0    3    0    0    3    0
</code></pre>
</div>

<p>Find all pattern matches in reference</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">myvpos</span> <span class="o">&lt;-</span> <span class="n">vmatchPattern</span><span class="p">(</span><span class="s2">"ATGGCT"</span><span class="p">,</span> <span class="n">myseq1</span><span class="p">,</span> <span class="n">max.mismatch</span><span class="o">=</span><span class="m">1</span><span class="p">)</span> 
<span class="n">myvpos</span> <span class="c1"># The results are stored as MIndex object.
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## MIndex object of length 2058
## $`gi|12057215|gb|AE004437.1|:248-1453 Halobacterium sp. NRC-1, complete genome`
## IRanges of length 3
##     start end width
## [1]     1   6     6
## [2]   383 388     6
## [3]   928 933     6
## 
## $`gi|12057215|gb|AE004437.1|:1450-2115 Halobacterium sp. NRC-1, complete genome`
## IRanges of length 0
## 
## $`gi|12057215|gb|AE004437.1|:2145-3254 Halobacterium sp. NRC-1, complete genome`
## IRanges of length 5
##     start end width
## [1]     1   6     6
## [2]    94  99     6
## [3]   221 226     6
## [4]   535 540     6
## [5]   601 606     6
## 
## ...
## &lt;2055 more elements&gt;
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">Views</span><span class="p">(</span><span class="n">myseq1</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span> <span class="n">start</span><span class="p">(</span><span class="n">myvpos</span><span class="p">[[</span><span class="m">1</span><span class="p">]]),</span> <span class="n">end</span><span class="p">(</span><span class="n">myvpos</span><span class="p">[[</span><span class="m">1</span><span class="p">]]))</span> <span class="c1"># Retrieves the result for single entry
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Views on a 1206-letter DNAString subject
## subject: ATGACTCGGCGGTCTCGTGTCGGTGCCGGCCTCGCAGCCATTGT...TTGCGATCGTCGTCGTCGTTGTTCGACGCTGGCGGAACCCATGA
## views:
##     start end width
## [1]     1   6     6 [ATGACT]
## [2]   383 388     6 [ATGGCA]
## [3]   928 933     6 [ATGACT]
</code></pre>
</div>

<p>Return all matches</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">myseq1</span><span class="p">),</span> <span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> 
       <span class="n">as.character</span><span class="p">(</span><span class="n">Views</span><span class="p">(</span><span class="n">myseq1</span><span class="p">[[</span><span class="n">x</span><span class="p">]],</span> <span class="n">start</span><span class="p">(</span><span class="n">myvpos</span><span class="p">[[</span><span class="n">x</span><span class="p">]]),</span> <span class="n">end</span><span class="p">(</span><span class="n">myvpos</span><span class="p">[[</span><span class="n">x</span><span class="p">]]))))[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">]</span> 
</code></pre>
</div>

<h3 id="pattern-matching-with-regular-expression-support">Pattern matching with regular expression support</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">myseq</span> <span class="o">&lt;-</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"ATGCAGACATAGTG"</span><span class="p">,</span> <span class="s2">"ATGAACATAGATCC"</span><span class="p">,</span> <span class="s2">"GTACAGATCAC"</span><span class="p">))</span>
<span class="n">myseq</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"^ATG"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">,</span> <span class="n">perl</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)]</span> <span class="c1"># String searching with regular expression support
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]    14 ATGCAGACATAGTG
## [2]    14 ATGAACATAGATCC
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">pos1</span> <span class="o">&lt;-</span> <span class="n">regexpr</span><span class="p">(</span><span class="s2">"AT"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">)</span> <span class="c1"># Searches 'myseq' for first match of pattern "AT"
</span><span class="n">as.numeric</span><span class="p">(</span><span class="n">pos1</span><span class="p">);</span> <span class="n">attributes</span><span class="p">(</span><span class="n">pos1</span><span class="p">)</span><span class="o">$</span><span class="n">match.length</span> <span class="c1"># Returns position information of matches
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 1 7
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 2 2 2
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">pos2</span> <span class="o">&lt;-</span> <span class="n">gregexpr</span><span class="p">(</span><span class="s2">"AT"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">)</span> <span class="c1"># Searches 'myseq' for all matches of pattern "AT"
</span><span class="n">as.numeric</span><span class="p">(</span><span class="n">pos2</span><span class="p">[[</span><span class="m">1</span><span class="p">]]);</span> <span class="n">attributes</span><span class="p">(</span><span class="n">pos2</span><span class="p">[[</span><span class="m">1</span><span class="p">]])</span><span class="o">$</span><span class="n">match.length</span> <span class="c1"># Match positions in first sequence
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 9
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 2 2
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">DNAStringSet</span><span class="p">(</span><span class="n">gsub</span><span class="p">(</span><span class="s2">"^ATG"</span><span class="p">,</span> <span class="s2">"NNN"</span><span class="p">,</span> <span class="n">myseq</span><span class="p">))</span> <span class="c1"># String substitution with regular expression support
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 3
##     width seq
## [1]    14 NNNCAGACATAGTG
## [2]    14 NNNAACATAGATCC
## [3]    11 GTACAGATCAC
</code></pre>
</div>

<h2 id="pwm-viewing-and-searching">PWM Viewing and Searching</h2>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">seqLogo</span><span class="p">)</span> 
<span class="n">pwm</span> <span class="o">&lt;-</span> <span class="n">PWM</span><span class="p">(</span><span class="n">DNAStringSet</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"GCT"</span><span class="p">,</span> <span class="s2">"GGT"</span><span class="p">,</span> <span class="s2">"GCA"</span><span class="p">)))</span> 
<span class="n">pwm</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##        [,1]      [,2]      [,3]
## A 0.0000000 0.0000000 0.2312611
## C 0.0000000 0.3157205 0.0000000
## G 0.3685591 0.2312611 0.0000000
## T 0.0000000 0.0000000 0.3157205
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">seqLogo</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">pwm</span><span class="p">)</span> <span class="o">*</span> <span class="m">1</span><span class="o">/</span><span class="n">colSums</span><span class="p">(</span><span class="n">pwm</span><span class="p">)))</span>
</code></pre>
</div>

<p><img src="Rsequences_files/figure-html/pwm_logo-1.png" alt="" />\</p>

<p>Search sequence for PWM matches with score better than <code class="highlighter-rouge">min.score</code></p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">DNAString</span><span class="p">(</span><span class="s2">"AAAGCTAAAGGTAAAGCAAAA"</span><span class="p">)</span> 
<span class="n">matchPWM</span><span class="p">(</span><span class="n">pwm</span><span class="p">,</span> <span class="n">chr</span><span class="p">,</span> <span class="n">min.score</span><span class="o">=</span><span class="m">0.9</span><span class="p">)</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   Views on a 21-letter DNAString subject
## subject: AAAGCTAAAGGTAAAGCAAAA
## views:
##     start end width
## [1]     4   6     3 [GCT]
## [2]    10  12     3 [GGT]
## [3]    16  18     3 [GCA]
</code></pre>
</div>

<h1 id="ngs-sequences">NGS Sequences</h1>

<h2 id="sequence-and-quality-data-fastq-format">Sequence and Quality Data: FASTQ Format</h2>

<p>Four lines per sequence:</p>

<ol>
  <li>ID</li>
  <li>Sequence</li>
  <li>ID</li>
  <li>Base call qualities (Phred scores) as ASCII characters</li>
</ol>

<p>The following gives an example of 3 Illumina reads in a FASTQ file. The numbers
at the beginning of each line are not part of the FASTQ format. They have been added 
solely for illustration purposes.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>1. @SRR038845.3 HWI-EAS038:6:1:0:1938 length=36
2. CAACGAGTTCACACCTTGGCCGACAGGCCCGGGTAA
3. +SRR038845.3 HWI-EAS038:6:1:0:1938 length=36
4. BA@7&gt;B=&gt;:&gt;&gt;7@7@&gt;&gt;9=BAA?;&gt;52;&gt;:9=8.=A
1. @SRR038845.41 HWI-EAS038:6:1:0:1474 length=36
2. CCAATGATTTTTTTCCGTGTTTCAGAATACGGTTAA
3. +SRR038845.41 HWI-EAS038:6:1:0:1474 length=36
4. BCCBA@BB@BBBBAB@B9B@=BABA@A:@693:@B=
1. @SRR038845.53 HWI-EAS038:6:1:1:360 length=36
2. GTTCAAAAAGAACTAAATTGTGTCAATAGAAAACTC
3. +SRR038845.53 HWI-EAS038:6:1:1:360 length=36
4. BBCBBBBBB@@BAB?BBBBCBC&gt;BBBAA8&gt;BBBAA@
</code></pre>
</div>

<h2 id="sequence-and-quality-data-qualityscalexstringset">Sequence and Quality Data: <code class="highlighter-rouge">QualityScaleXStringSet</code></h2>

<p>Phred quality scores are integers from 0-50 that are
stored as ASCII characters after adding 33. The basic R functions <code class="highlighter-rouge">rawToChar</code> and
<code class="highlighter-rouge">charToRaw</code> can be used to interconvert among their representations.</p>

<p>Phred score interconversion</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">phred</span> <span class="o">&lt;-</span> <span class="m">1</span><span class="o">:</span><span class="m">9</span>
<span class="n">phreda</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">as.raw</span><span class="p">((</span><span class="n">phred</span><span class="p">)</span><span class="m">+33</span><span class="p">),</span> <span class="n">rawToChar</span><span class="p">),</span> <span class="n">collapse</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span>
<span class="n">phreda</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] "\"#$%&amp;'()*"
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">as.integer</span><span class="p">(</span><span class="n">charToRaw</span><span class="p">(</span><span class="n">phreda</span><span class="p">))</span><span class="m">-33</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 1 2 3 4 5 6 7 8 9
</code></pre>
</div>

<p>Construct <code class="highlighter-rouge">QualityScaledDNAStringSet</code> from scratch</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">dset</span> <span class="o">&lt;-</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">,</span> <span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">paste</span><span class="p">(</span><span class="n">sample</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"A"</span><span class="p">,</span><span class="s2">"T"</span><span class="p">,</span><span class="s2">"G"</span><span class="p">,</span><span class="s2">"C"</span><span class="p">),</span> <span class="m">20</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="n">T</span><span class="p">),</span> <span class="n">collapse</span><span class="o">=</span><span class="s2">""</span><span class="p">)))</span> <span class="c1"># Creates random sample sequence.
</span><span class="n">myqlist</span> <span class="o">&lt;-</span> <span class="n">lapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">100</span><span class="p">,</span> <span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">sample</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">40</span><span class="p">,</span> <span class="m">20</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="n">T</span><span class="p">))</span> <span class="c1"># Creates random Phred score list.
</span><span class="n">myqual</span> <span class="o">&lt;-</span> <span class="n">sapply</span><span class="p">(</span><span class="n">myqlist</span><span class="p">,</span> <span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">toString</span><span class="p">(</span><span class="n">PhredQuality</span><span class="p">(</span><span class="n">x</span><span class="p">)))</span> <span class="c1"># Converts integer scores into ASCII characters.
</span><span class="n">myqual</span> <span class="o">&lt;-</span> <span class="n">PhredQuality</span><span class="p">(</span><span class="n">myqual</span><span class="p">)</span> <span class="c1"># Converts to a PhredQuality object.
</span><span class="n">dsetq1</span> <span class="o">&lt;-</span> <span class="n">QualityScaledDNAStringSet</span><span class="p">(</span><span class="n">dset</span><span class="p">,</span> <span class="n">myqual</span><span class="p">)</span> <span class="c1"># Combines DNAStringSet and quality data in QualityScaledDNAStringSet object.
</span><span class="n">dsetq1</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A QualityScaledDNAStringSet instance containing:
## 
##   A DNAStringSet instance of length 2
##     width seq
## [1]    20 TGAGGTCTGGCCCATTGAAT
## [2]    20 ACGGCGGTAAGTCCTGCAAA
## 
##   A PhredQuality instance of length 2
##     width seq
## [1]    20 ;;A1HH/"@2E/2D"9/(DD
## [2]    20 A2+7.IG:-&gt;0*3I7"5H-%
</code></pre>
</div>

<h2 id="processing-fastq-files-with-shortread">Processing FASTQ Files with ShortRead</h2>

<p>The following expains the basic usage of <code class="highlighter-rouge">ShortReadQ</code> objects. To make the sample code work, 
download and unzip this <a href="http://faculty.ucr.edu/~tgirke/HTML_Presentations/Manuals/Workshop_Dec_6_10_2012/Rsequences/data.zip">file</a> to your current working directory.
The following code performs the download for you.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ShortRead</span><span class="p">)</span>
<span class="n">download.file</span><span class="p">(</span><span class="s2">"http://faculty.ucr.edu/~tgirke/HTML_Presentations/Manuals/Workshop_Dec_6_10_2012/Rsequences/data.zip"</span><span class="p">,</span> <span class="s2">"data.zip"</span><span class="p">)</span>
<span class="n">unzip</span><span class="p">(</span><span class="s2">"data.zip"</span><span class="p">)</span>
</code></pre>
</div>

<p>Important utilities for accessing FASTQ files</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">fastq</span> <span class="o">&lt;-</span> <span class="n">list.files</span><span class="p">(</span><span class="s2">"data"</span><span class="p">,</span> <span class="s2">"*.fastq$"</span><span class="p">);</span> <span class="n">fastq</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="s2">"data/"</span><span class="p">,</span> <span class="n">fastq</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span>
<span class="n">names</span><span class="p">(</span><span class="n">fastq</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="s2">"flowcell6_lane"</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="n">length</span><span class="p">(</span><span class="n">fastq</span><span class="p">),</span> <span class="n">sep</span><span class="o">=</span><span class="s2">"_"</span><span class="p">)</span> 
<span class="p">(</span><span class="n">fq</span> <span class="o">&lt;-</span> <span class="n">readFastq</span><span class="p">(</span><span class="n">fastq</span><span class="p">[</span><span class="m">1</span><span class="p">]))</span> <span class="c1"># Imports first FASTQ file
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## class: ShortReadQ
## length: 1000 reads; width: 36 cycles
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">countLines</span><span class="p">(</span><span class="n">dirPath</span><span class="o">=</span><span class="s2">"./data"</span><span class="p">,</span> <span class="n">pattern</span><span class="o">=</span><span class="s2">".fastq$"</span><span class="p">)</span><span class="o">/</span><span class="m">4</span> <span class="c1"># Counts numbers of reads in FASTQ files
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## SRR038845.fastq SRR038846.fastq SRR038848.fastq SRR038850.fastq 
##            1000            1000            1000            1000
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">id</span><span class="p">(</span><span class="n">fq</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span> <span class="c1"># Returns ID field
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A BStringSet instance of length 1
##     width seq
## [1]    43 SRR038845.3 HWI-EAS038:6:1:0:1938 length=36
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sread</span><span class="p">(</span><span class="n">fq</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span> <span class="c1"># Returns sequence
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 1
##     width seq
## [1]    36 CAACGAGTTCACACCTTGGCCGACAGGCCCGGGTAA
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">quality</span><span class="p">(</span><span class="n">fq</span><span class="p">)[</span><span class="m">1</span><span class="p">]</span> <span class="c1"># Returns Phred scores 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## class: FastqQuality
## quality:
##   A BStringSet instance of length 1
##     width seq
## [1]    36 BA@7&gt;B=&gt;:&gt;&gt;7@7@&gt;&gt;9=BAA?;&gt;52;&gt;:9=8.=A
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">as</span><span class="p">(</span><span class="n">quality</span><span class="p">(</span><span class="n">fq</span><span class="p">),</span> <span class="s2">"matrix"</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">12</span><span class="p">]</span> <span class="c1"># Coerces Phred scores to numeric matrix
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]
## [1,]   33   32   31   22   29   33   28   29   25    29    29    22
## [2,]   33   34   34   33   32   31   33   33   31    33    33    33
## [3,]   33   33   34   33   33   33   33   33   33    31    31    33
## [4,]   33   33   33   33   31   33   28   31   28    32    33    33
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">ShortReadQ</span><span class="p">(</span><span class="n">sread</span><span class="o">=</span><span class="n">sread</span><span class="p">(</span><span class="n">fq</span><span class="p">),</span> <span class="n">quality</span><span class="o">=</span><span class="n">quality</span><span class="p">(</span><span class="n">fq</span><span class="p">),</span> <span class="n">id</span><span class="o">=</span><span class="n">id</span><span class="p">(</span><span class="n">fq</span><span class="p">))</span> <span class="c1"># Constructs a ShortReadQ from components
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## class: ShortReadQ
## length: 1000 reads; width: 36 cycles
</code></pre>
</div>

<h2 id="fastq-quality-reports">FASTQ Quality Reports</h2>

<h3 id="using-systempiper">Using <code class="highlighter-rouge">systemPipeR</code></h3>

<p>The following <code class="highlighter-rouge">seeFastq</code> and <code class="highlighter-rouge">seeFastqPlot</code> functions generate and plot a series of useful quality statistics for a set of FASTQ files.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span>
<span class="n">fqlist</span> <span class="o">&lt;-</span> <span class="n">seeFastq</span><span class="p">(</span><span class="n">fastq</span><span class="o">=</span><span class="n">fastq</span><span class="p">,</span> <span class="n">batchsize</span><span class="o">=</span><span class="m">800</span><span class="p">,</span> <span class="n">klength</span><span class="o">=</span><span class="m">8</span><span class="p">)</span> <span class="c1"># For real data set batchsize to at least 10^5 
</span><span class="n">seeFastqPlot</span><span class="p">(</span><span class="n">fqlist</span><span class="p">)</span>
</code></pre>
</div>

<p><img src="Rsequences_files/figure-html/see_fastq-1.png" alt="" />\</p>

<p>Handles many samples in one PDF file. For more details see <a href="http://bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html">here</a></p>

<h3 id="using-shortread">Using <code class="highlighter-rouge">ShortRead</code></h3>

<p>The <code class="highlighter-rouge">ShortRead</code> package contains several FASTQ quality reporting functions.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sp</span> <span class="o">&lt;-</span> <span class="n">SolexaPath</span><span class="p">(</span><span class="n">system.file</span><span class="p">(</span><span class="s1">'extdata'</span><span class="p">,</span> <span class="n">package</span><span class="o">=</span><span class="s1">'ShortRead'</span><span class="p">))</span>
<span class="n">fl</span> <span class="o">&lt;-</span> <span class="n">file.path</span><span class="p">(</span><span class="n">analysisPath</span><span class="p">(</span><span class="n">sp</span><span class="p">),</span> <span class="s2">"s_1_sequence.txt"</span><span class="p">)</span> 
<span class="n">fls</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">fl</span><span class="p">,</span> <span class="n">fl</span><span class="p">)</span> 
<span class="n">coll</span> <span class="o">&lt;-</span> <span class="n">QACollate</span><span class="p">(</span><span class="n">QAFastqSource</span><span class="p">(</span><span class="n">fls</span><span class="p">),</span> <span class="n">QAReadQuality</span><span class="p">(),</span> <span class="n">QAAdapterContamination</span><span class="p">(),</span> 
	    <span class="n">QANucleotideUse</span><span class="p">(),</span> <span class="n">QAQualityUse</span><span class="p">(),</span> <span class="n">QASequenceUse</span><span class="p">(),</span> <span class="n">QAFrequentSequence</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="m">10</span><span class="p">),</span> 
		<span class="n">QANucleotideByCycle</span><span class="p">(),</span> <span class="n">QAQualityByCycle</span><span class="p">())</span>
<span class="n">x</span> <span class="o">&lt;-</span> <span class="n">qa2</span><span class="p">(</span><span class="n">coll</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span>
<span class="n">res</span> <span class="o">&lt;-</span> <span class="n">report</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="k">if</span><span class="p">(</span><span class="n">interactive</span><span class="p">())</span>
<span class="n">browseURL</span><span class="p">(</span><span class="n">res</span><span class="p">)</span> 
</code></pre>
</div>

<h2 id="filtering-and-trimming-fastq-files-with-shortread">Filtering and Trimming FASTQ Files with ShortRead</h2>

<h3 id="adaptor-trimming">Adaptor trimming</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">fqtrim</span> <span class="o">&lt;-</span> <span class="n">trimLRPatterns</span><span class="p">(</span><span class="n">Rpattern</span><span class="o">=</span><span class="s2">"GCCCGGGTAA"</span><span class="p">,</span> <span class="n">subject</span><span class="o">=</span><span class="n">fq</span><span class="p">)</span>
<span class="n">sread</span><span class="p">(</span><span class="n">fq</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">]</span> <span class="c1"># Before trimming
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]    36 CAACGAGTTCACACCTTGGCCGACAGGCCCGGGTAA
## [2]    36 CCAATGATTTTTTTCCGTGTTTCAGAATACGGTTAA
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sread</span><span class="p">(</span><span class="n">fqtrim</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">]</span> <span class="c1"># After trimming
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]    26 CAACGAGTTCACACCTTGGCCGACAG
## [2]    36 CCAATGATTTTTTTCCGTGTTTCAGAATACGGTTAA
</code></pre>
</div>
<p>### Read counting and duplicate removal</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">tables</span><span class="p">(</span><span class="n">fq</span><span class="p">)</span><span class="o">$</span><span class="n">distribution</span> <span class="c1"># Counts read occurences
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   nOccurrences nReads
## 1            1    948
## 2            2     26
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sum</span><span class="p">(</span><span class="n">srduplicated</span><span class="p">(</span><span class="n">fq</span><span class="p">))</span> <span class="c1"># Identifies duplicated reads
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 26
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">fq</span><span class="p">[</span><span class="o">!</span><span class="n">srduplicated</span><span class="p">(</span><span class="n">fq</span><span class="p">)]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## class: ShortReadQ
## length: 974 reads; width: 36 cycles
</code></pre>
</div>

<h3 id="trimming-low-quality-tails">Trimming low quality tails</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">cutoff</span> <span class="o">&lt;-</span> <span class="m">30</span>
<span class="n">cutoff</span> <span class="o">&lt;-</span> <span class="n">rawToChar</span><span class="p">(</span><span class="n">as.raw</span><span class="p">(</span><span class="n">cutoff</span><span class="m">+33</span><span class="p">))</span>
<span class="n">sread</span><span class="p">(</span><span class="n">trimTails</span><span class="p">(</span><span class="n">fq</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">a</span><span class="o">=</span><span class="n">cutoff</span><span class="p">,</span> <span class="n">successive</span><span class="o">=</span><span class="n">FALSE</span><span class="p">))[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 2
##     width seq
## [1]     4 CAAC
## [2]    20 CCAATGATTTTTTTCCGTGT
</code></pre>
</div>

<h3 id="removal-of-reads-with-phred-scores-below-a-threshold-value">Removal of reads with Phred scores below a threshold value</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">cutoff</span> <span class="o">&lt;-</span> <span class="m">30</span>
<span class="n">qcount</span> <span class="o">&lt;-</span> <span class="n">rowSums</span><span class="p">(</span><span class="n">as</span><span class="p">(</span><span class="n">quality</span><span class="p">(</span><span class="n">fq</span><span class="p">),</span> <span class="s2">"matrix"</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="m">20</span><span class="p">)</span> 
<span class="n">fq</span><span class="p">[</span><span class="n">qcount</span> <span class="o">==</span> <span class="m">0</span><span class="p">]</span> <span class="c1"># Number of reads where all Phred scores &gt;= 20
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## class: ShortReadQ
## length: 349 reads; width: 36 cycles
</code></pre>
</div>

<h3 id="removal-of-reads-with-x-ns-andor-low-complexity-segments">Removal of reads with x Ns and/or low complexity segments</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">filter1</span> <span class="o">&lt;-</span> <span class="n">nFilter</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="m">1</span><span class="p">)</span> <span class="c1"># Keeps only reads without Ns
</span><span class="n">filter2</span> <span class="o">&lt;-</span> <span class="n">polynFilter</span><span class="p">(</span><span class="n">threshold</span><span class="o">=</span><span class="m">20</span><span class="p">,</span> <span class="n">nuc</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">"A"</span><span class="p">,</span><span class="s2">"T"</span><span class="p">,</span><span class="s2">"G"</span><span class="p">,</span><span class="s2">"C"</span><span class="p">))</span> <span class="c1"># Removes reads with &gt;=20 of one nucleotide
</span><span class="n">filter</span> <span class="o">&lt;-</span> <span class="n">compose</span><span class="p">(</span><span class="n">filter1</span><span class="p">,</span> <span class="n">filter2</span><span class="p">)</span>
<span class="n">fq</span><span class="p">[</span><span class="n">filter</span><span class="p">(</span><span class="n">fq</span><span class="p">)]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## class: ShortReadQ
## length: 989 reads; width: 36 cycles
</code></pre>
</div>

<h2 id="memory-efficient-fastq-processing">Memory Efficient FASTQ Processing</h2>

<p>Streaming through FASTQ files with <code class="highlighter-rouge">FastqStreamer</code> and random sampling reads with <code class="highlighter-rouge">FastqSampler</code></p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">fq</span> <span class="o">&lt;-</span> <span class="n">yield</span><span class="p">(</span><span class="n">FastqStreamer</span><span class="p">(</span><span class="n">fastq</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="m">50</span><span class="p">))</span> <span class="c1"># Imports first 50 reads 
</span><span class="n">fq</span> <span class="o">&lt;-</span> <span class="n">yield</span><span class="p">(</span><span class="n">FastqSampler</span><span class="p">(</span><span class="n">fastq</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="m">50</span><span class="p">))</span> <span class="c1"># Random samples 50 reads 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning: closing unused connection 5 (data/SRR038845.fastq)
</code></pre>
</div>

<p>Streaming through a FASTQ file while applying filtering/trimming functions and writing the results to a new file
 here <code class="highlighter-rouge">SRR038845.fastq_sub</code> in <code class="highlighter-rouge">data</code> directory.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">f</span> <span class="o">&lt;-</span> <span class="n">FastqStreamer</span><span class="p">(</span><span class="n">fastq</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="m">50</span><span class="p">)</span> 
<span class="k">while</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">fq</span> <span class="o">&lt;-</span> <span class="n">yield</span><span class="p">(</span><span class="n">f</span><span class="p">)))</span> <span class="p">{</span>
	<span class="n">fqsub</span> <span class="o">&lt;-</span> <span class="n">fq</span><span class="p">[</span><span class="n">grepl</span><span class="p">(</span><span class="s2">"^TT"</span><span class="p">,</span> <span class="n">sread</span><span class="p">(</span><span class="n">fq</span><span class="p">))]</span> 
	<span class="n">writeFastq</span><span class="p">(</span><span class="n">fqsub</span><span class="p">,</span> <span class="n">paste</span><span class="p">(</span><span class="n">fastq</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="s2">"sub"</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">"_"</span><span class="p">),</span> <span class="n">mode</span><span class="o">=</span><span class="s2">"a"</span><span class="p">,</span> <span class="n">compress</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">close</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
</code></pre>
</div>

<h1 id="range-operations">Range Operations</h1>

<h2 id="important-data-objects-for-range-operations">Important Data Objects for Range Operations</h2>

<ul>
  <li><code class="highlighter-rouge">IRanges</code>: stores range data only (IRanges library)</li>
  <li><code class="highlighter-rouge">GRanges</code>: stores ranges and annotations (GenomicRanges library)</li>
  <li><code class="highlighter-rouge">GRangesList</code>: list version of GRanges container (GenomicRanges library)</li>
</ul>

<h2 id="range-data-are-stored-in-iranges-and-granges-containers">Range Data Are Stored in <code class="highlighter-rouge">IRanges</code> and <code class="highlighter-rouge">GRanges</code> Containers</h2>

<h3 id="construct-granges-object">Construct <code class="highlighter-rouge">GRanges</code> Object</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">GenomicRanges</span><span class="p">);</span> <span class="n">library</span><span class="p">(</span><span class="n">rtracklayer</span><span class="p">)</span>
<span class="n">gr</span> <span class="o">&lt;-</span> <span class="n">GRanges</span><span class="p">(</span><span class="n">seqnames</span> <span class="o">=</span> <span class="n">Rle</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"chr1"</span><span class="p">,</span> <span class="s2">"chr2"</span><span class="p">,</span> <span class="s2">"chr1"</span><span class="p">,</span> <span class="s2">"chr3"</span><span class="p">),</span> <span class="n">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="m">4</span><span class="p">)),</span> <span class="n">ranges</span> <span class="o">=</span> <span class="n">IRanges</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="n">end</span> <span class="o">=</span> <span class="m">7</span><span class="o">:</span><span class="m">16</span><span class="p">,</span> <span class="n">names</span> <span class="o">=</span> <span class="n">head</span><span class="p">(</span><span class="n">letters</span><span class="p">,</span> <span class="m">10</span><span class="p">)),</span> <span class="n">strand</span> <span class="o">=</span> <span class="n">Rle</span><span class="p">(</span><span class="n">strand</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"-"</span><span class="p">,</span> <span class="s2">"+"</span><span class="p">,</span> <span class="s2">"*"</span><span class="p">,</span> <span class="s2">"+"</span><span class="p">,</span> <span class="s2">"-"</span><span class="p">)),</span> <span class="n">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span> <span class="m">3</span><span class="p">,</span> <span class="m">2</span><span class="p">)),</span> <span class="n">score</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span> <span class="n">GC</span> <span class="o">=</span> <span class="n">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="n">length</span> <span class="o">=</span> <span class="m">10</span><span class="p">))</span> <span class="c1"># Example of creating a GRanges object with its constructor function.
</span></code></pre>
</div>

<h3 id="import-gff-into-granges-object">Import GFF into <code class="highlighter-rouge">GRanges</code> Object</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span> <span class="o">&lt;-</span> <span class="n">import.gff</span><span class="p">(</span><span class="s2">"http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/gff3.gff"</span><span class="p">)</span> <span class="c1"># Imports a simplified GFF3 genome annotation file.
</span><span class="n">seqlengths</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">end</span><span class="p">(</span><span class="n">ranges</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)[,</span><span class="s2">"type"</span><span class="p">]</span><span class="o">==</span><span class="s2">"chromosome"</span><span class="p">),]))</span> 
<span class="n">names</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="m">1</span><span class="o">:</span><span class="n">length</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> <span class="c1"># Assigns names to corresponding slot
</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 10 metadata columns:
##     seqnames           ranges strand |   source       type     score     phase                  ID
##        &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##   1     Chr1 [   1, 30427671]      + |   TAIR10 chromosome      &lt;NA&gt;      &lt;NA&gt;                Chr1
##   2     Chr1 [3631,     5899]      + |   TAIR10       gene      &lt;NA&gt;      &lt;NA&gt;           AT1G01010
##   3     Chr1 [3631,     5899]      + |   TAIR10       mRNA      &lt;NA&gt;      &lt;NA&gt;         AT1G01010.1
##   4     Chr1 [3760,     5630]      + |   TAIR10    protein      &lt;NA&gt;      &lt;NA&gt; AT1G01010.1-Protein
##            Name                Note          Parent       Index Derives_from
##     &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##   1        Chr1                                            &lt;NA&gt;         &lt;NA&gt;
##   2   AT1G01010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   3 AT1G01010.1                           AT1G01010           1         &lt;NA&gt;
##   4 AT1G01010.1                                            &lt;NA&gt;  AT1G01010.1
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<h3 id="coerce-granges-object-to-dataframe">Coerce <code class="highlighter-rouge">GRanges</code> object to <code class="highlighter-rouge">data.frame</code></h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">as.data.frame</span><span class="p">(</span><span class="n">gff</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span> <span class="m">1</span><span class="o">:</span><span class="m">7</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   seqnames start      end    width strand source       type
## 1     Chr1     1 30427671 30427671      + TAIR10 chromosome
## 2     Chr1  3631     5899     2269      + TAIR10       gene
## 3     Chr1  3631     5899     2269      + TAIR10       mRNA
## 4     Chr1  3760     5630     1871      + TAIR10    protein
</code></pre>
</div>

<h3 id="coerce-granges-to-rangeddata-object-and-vice-versa">Coerce <code class="highlighter-rouge">GRanges</code> to <code class="highlighter-rouge">RangedData</code> object and vice versa</h3>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff_rd</span> <span class="o">&lt;-</span> <span class="n">as</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span> <span class="s2">"RangedData"</span><span class="p">)</span> 
<span class="n">gff_gr</span> <span class="o">&lt;-</span> <span class="n">as</span><span class="p">(</span><span class="n">gff_rd</span><span class="p">,</span> <span class="s2">"GRanges"</span><span class="p">)</span> 
</code></pre>
</div>

<h2 id="utilities-for-range-containers">Utilities for Range Containers</h2>

<h3 id="accessor-and-subsetting-methods-for-granges-objects">Accessor and subsetting methods for GRanges objects</h3>

<p>Subsetting and replacement</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 10 metadata columns:
##     seqnames           ranges strand |   source       type     score     phase                  ID
##        &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##   1     Chr1 [   1, 30427671]      + |   TAIR10 chromosome      &lt;NA&gt;      &lt;NA&gt;                Chr1
##   2     Chr1 [3631,     5899]      + |   TAIR10       gene      &lt;NA&gt;      &lt;NA&gt;           AT1G01010
##   3     Chr1 [3631,     5899]      + |   TAIR10       mRNA      &lt;NA&gt;      &lt;NA&gt;         AT1G01010.1
##   4     Chr1 [3760,     5630]      + |   TAIR10    protein      &lt;NA&gt;      &lt;NA&gt; AT1G01010.1-Protein
##            Name                Note          Parent       Index Derives_from
##     &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##   1        Chr1                                            &lt;NA&gt;         &lt;NA&gt;
##   2   AT1G01010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   3 AT1G01010.1                           AT1G01010           1         &lt;NA&gt;
##   4 AT1G01010.1                                            &lt;NA&gt;  AT1G01010.1
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="s2">"type"</span><span class="p">,</span> <span class="s2">"ID"</span><span class="p">)]</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 2 metadata columns:
##     seqnames           ranges strand |       type                  ID
##        &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;factor&gt;         &lt;character&gt;
##   1     Chr1 [   1, 30427671]      + | chromosome                Chr1
##   2     Chr1 [3631,     5899]      + |       gene           AT1G01010
##   3     Chr1 [3631,     5899]      + |       mRNA         AT1G01010.1
##   4     Chr1 [3760,     5630]      + |    protein AT1G01010.1-Protein
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="m">2</span><span class="p">]</span> <span class="o">&lt;-</span> <span class="n">gff</span><span class="p">[</span><span class="m">3</span><span class="p">]</span> 
</code></pre>
</div>

<p>GRanges objects can be concatenated with the <code class="highlighter-rouge">c</code> function</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">c</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span> <span class="n">gff</span><span class="p">[</span><span class="m">401</span><span class="o">:</span><span class="m">402</span><span class="p">])</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 10 metadata columns:
##       seqnames           ranges strand |   source           type     score     phase
##          &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##     1     Chr1 [   1, 30427671]      + |   TAIR10     chromosome      &lt;NA&gt;      &lt;NA&gt;
##     2     Chr1 [3631,     5899]      + |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##   401     Chr5 [5516,     5769]      - |   TAIR10        protein      &lt;NA&gt;      &lt;NA&gt;
##   402     Chr5 [5770,     5801]      - |   TAIR10 five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##                        ID        Name            Note          Parent       Index Derives_from
##               &lt;character&gt; &lt;character&gt; &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##     1                Chr1        Chr1                                        &lt;NA&gt;         &lt;NA&gt;
##     2         AT1G01010.1 AT1G01010.1                       AT1G01010           1         &lt;NA&gt;
##   401 AT5G01015.2-Protein AT5G01015.2                                        &lt;NA&gt;  AT5G01015.2
##   402                &lt;NA&gt;        &lt;NA&gt;                     AT5G01015.2        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Acessor functions</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">seqnames</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## factor-Rle of length 449 with 7 runs
##   Lengths:   72   22   38  118  172   13   14
##   Values : Chr1 Chr2 Chr3 Chr4 Chr5 ChrC ChrM
## Levels(7): Chr1 Chr2 Chr3 Chr4 Chr5 ChrC ChrM
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">ranges</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## IRanges of length 449
##       start      end    width names
## [1]       1 30427671 30427671     1
## [2]    3631     5899     2269     2
## [3]    3631     5899     2269     3
## [4]    3760     5630     1871     4
## [5]    3631     3913      283     5
## ...     ...      ...      ...   ...
## [445] 11918    12241      324   445
## [446] 11918    12241      324   446
## [447] 11918    12241      324   447
## [448] 11918    12241      324   448
## [449] 11918    12241      324   449
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">strand</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## factor-Rle of length 449 with 13 runs
##   Lengths:  18  54  28  21  12 117   1 171   1  12   1   8   5
##   Values :   +   -   +   -   +   -   +   -   +   -   +   -   +
## Levels(3): + - *
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">seqlengths</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     Chr1     Chr2     Chr3     Chr4     Chr5     ChrC     ChrM 
## 30427671 19698289 23459830 18585056 26975502   154478   366924
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">start</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1]    1 3631 3631 3760
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">end</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 30427671     5899     5899     5630
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">width</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 30427671     2269     2269     1871
</code></pre>
</div>

<p>Accessing metadata component</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> <span class="c1"># or elementMetadata(gff)
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## DataFrame with 449 rows and 10 columns
##       source       type     score     phase                  ID        Name                Note
##     &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt; &lt;character&gt;     &lt;CharacterList&gt;
## 1     TAIR10 chromosome        NA        NA                Chr1        Chr1                    
## 2     TAIR10       mRNA        NA        NA         AT1G01010.1 AT1G01010.1                    
## 3     TAIR10       mRNA        NA        NA         AT1G01010.1 AT1G01010.1                    
## 4     TAIR10    protein        NA        NA AT1G01010.1-Protein AT1G01010.1                    
## 5     TAIR10       exon        NA        NA                  NA          NA                    
## ...      ...        ...       ...       ...                 ...         ...                 ...
## 445   TAIR10       gene        NA        NA           ATMG00030   ATMG00030 protein_coding_gene
## 446   TAIR10       mRNA        NA        NA         ATMG00030.1 ATMG00030.1                    
## 447   TAIR10    protein        NA        NA ATMG00030.1-Protein ATMG00030.1                    
## 448   TAIR10       exon        NA        NA                  NA          NA                    
## 449   TAIR10        CDS        NA         0                  NA          NA                    
##                              Parent       Index Derives_from
##                     &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
## 1                                            NA           NA
## 2                         AT1G01010           1           NA
## 3                         AT1G01010           1           NA
## 4                                            NA  AT1G01010.1
## 5                       AT1G01010.1          NA           NA
## ...                             ...         ...          ...
## 445                                          NA           NA
## 446                       ATMG00030           1           NA
## 447                                          NA  ATMG00030.1
## 448                     ATMG00030.1          NA           NA
## 449 ATMG00030.1,ATMG00030.1-Protein          NA           NA
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)[,</span> <span class="s2">"type"</span><span class="p">][</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">]</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] chromosome      mRNA            mRNA            protein         exon            five_prime_UTR 
##  [7] CDS             exon            CDS             exon            CDS             exon           
## [13] CDS             exon            CDS             exon            CDS             three_prime_UTR
## [19] gene            mRNA           
## Levels: chromosome gene mRNA protein exon five_prime_UTR CDS three_prime_UTR rRNA tRNA
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="n">elementMetadata</span><span class="p">(</span><span class="n">gff</span><span class="p">)[</span> <span class="p">,</span><span class="s2">"type"</span><span class="p">]</span> <span class="o">==</span> <span class="s2">"gene"</span><span class="p">]</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 21 ranges and 10 metadata columns:
##       seqnames         ranges strand   |   source     type     score     phase          ID
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;
##    19     Chr1 [ 5928,  8737]      -   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT1G01020
##    64     Chr1 [11649, 13714]      -   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT1G01030
##    74     Chr2 [ 1025,  2810]      +   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT2G01008
##    84     Chr2 [ 3706,  5513]      +   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT2G01010
##    87     Chr2 [ 5782,  5945]      +   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT2G01020
##   ...      ...            ...    ... ...      ...      ...       ...       ...         ...
##   427     ChrC [  383,  1444]      -   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATCG00020
##   432     ChrC [ 1717,  4347]      -   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATCG00030
##   437     ChrM [  273,   734]      -   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATMG00010
##   442     ChrM [ 8848, 11415]      -   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATMG00020
##   445     ChrM [11918, 12241]      +   |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATMG00030
##              Name                Note          Parent       Index Derives_from
##       &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##    19   AT1G01020 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    64   AT1G01030 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    74   AT2G01008 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    84   AT2G01010                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##    87   AT2G01020                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##   ...         ...                 ...             ...         ...          ...
##   427   ATCG00020 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   432   ATCG00030                tRNA                        &lt;NA&gt;         &lt;NA&gt;
##   437   ATMG00010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   442   ATMG00020                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##   445   ATMG00030 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<h3 id="useful-utilities-for-granges-objects">Useful utilities for GRanges objects</h3>

<p>Remove chromosome ranges</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span> <span class="o">&lt;-</span> <span class="n">gff</span><span class="p">[</span><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="o">$</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">"chromosome"</span><span class="p">]</span> 
</code></pre>
</div>

<p>Erase the strand information</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">strand</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s2">"*"</span> 
</code></pre>
</div>

<p>Collapses overlapping ranges to continuous ranges.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">reduce</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 22 ranges and 0 metadata columns:
##        seqnames         ranges strand
##           &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1 [ 3631,  5899]      *
##    [2]     Chr1 [ 5928,  8737]      *
##    [3]     Chr1 [11649, 13714]      *
##    [4]     Chr2 [ 1025,  2810]      *
##    [5]     Chr2 [ 3706,  5513]      *
##    ...      ...            ...    ...
##   [18]     ChrC [  383,  1444]      *
##   [19]     ChrC [ 1717,  4347]      *
##   [20]     ChrM [  273,   734]      *
##   [21]     ChrM [ 8848, 11415]      *
##   [22]     ChrM [11918, 12241]      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Return uncovered regions</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gaps</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 43 ranges and 0 metadata columns:
##        seqnames           ranges strand
##           &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1 [   1, 30427671]      +
##    [2]     Chr1 [   1, 30427671]      -
##    [3]     Chr1 [   1,     3630]      *
##    [4]     Chr1 [5900,     5927]      *
##    [5]     Chr1 [8738,    11648]      *
##    ...      ...              ...    ...
##   [39]     ChrM  [    1, 366924]      -
##   [40]     ChrM  [    1,    272]      *
##   [41]     ChrM  [  735,   8847]      *
##   [42]     ChrM  [11416,  11917]      *
##   [43]     ChrM  [12242, 366924]      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Return disjoint ranges</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">disjoin</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 211 ranges and 0 metadata columns:
##         seqnames         ranges strand
##            &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;
##     [1]     Chr1   [3631, 3759]      *
##     [2]     Chr1   [3760, 3913]      *
##     [3]     Chr1   [3914, 3995]      *
##     [4]     Chr1   [3996, 4276]      *
##     [5]     Chr1   [4277, 4485]      *
##     ...      ...            ...    ...
##   [207]     ChrC [ 1752,  4310]      *
##   [208]     ChrC [ 4311,  4347]      *
##   [209]     ChrM [  273,   734]      *
##   [210]     ChrM [ 8848, 11415]      *
##   [211]     ChrM [11918, 12241]      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Returns coverage of ranges</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">coverage</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## RleList of length 7
## $Chr1
## integer-Rle of length 30427671 with 45 runs
##   Lengths:     3630      129      154       82      281 ...      233      161      380 30413957
##   Values :        0        4        5        3        5 ...        4        2        4        0
## 
## $Chr2
## integer-Rle of length 19698289 with 14 runs
##   Lengths:     1024      248      185       53      362 ...      164      625      102 19691617
##   Values :        0        5        3        5        3 ...        3        0        5        0
## 
## $Chr3
## integer-Rle of length 23459830 with 29 runs
##   Lengths:     1652      145      139      111       95 ...      155      148      156 23453781
##   Values :        0        4        5        3        5 ...        3        5        4        0
## 
## $Chr4
## integer-Rle of length 18585056 with 72 runs
##   Lengths:     1179      357     1358      128      872 ...      212      114       74 18571697
##   Values :        0        5        0        5        3 ...        3        5        4        0
## 
## $Chr5
## integer-Rle of length 26975502 with 64 runs
##   Lengths:     1222       28       28      109       72 ...       76       55      174 26967058
##   Values :        0        4        7       13       16 ...        3        5        4        0
## 
## ...
## &lt;2 more elements&gt;
</code></pre>
</div>

<p>Return the index pairings for overlapping ranges</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">findOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span> <span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Hits object with 55 hits and 0 metadata columns:
##        queryHits subjectHits
##        &lt;integer&gt;   &lt;integer&gt;
##    [1]         1           1
##    [2]         1           2
##    [3]         1           4
##    [4]         1           3
##    [5]         2           1
##    ...       ...         ...
##   [51]        16           1
##   [52]        16           2
##   [53]        16           3
##   [54]        17           1
##   [55]        17           2
##   -------
##   queryLength: 442
##   subjectLength: 4
</code></pre>
</div>

<p>Counts overlapping ranges</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">countOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span> <span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])[</span><span class="m">1</span><span class="o">:</span><span class="m">40</span><span class="p">]</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 
##  4  4  4  4  3  4  3  3  3  3  3  3  3  3  3  3  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 
## 35 36 37 38 39 40 41 
##  0  0  0  0  0  0  0
</code></pre>
</div>

<p>Return only overlapping ranges</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">subsetByOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span> <span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span> 
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 17 ranges and 10 metadata columns:
##       seqnames       ranges strand   |   source            type     score     phase
##          &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;factor&gt;        &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##     2     Chr1 [3631, 5899]      *   |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##     3     Chr1 [3631, 5899]      *   |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##     4     Chr1 [3760, 5630]      *   |   TAIR10         protein      &lt;NA&gt;      &lt;NA&gt;
##     5     Chr1 [3631, 3913]      *   |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##     6     Chr1 [3631, 3759]      *   |   TAIR10  five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##   ...      ...          ...    ... ...      ...             ...       ...       ...
##    14     Chr1 [5174, 5326]      *   |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##    15     Chr1 [5174, 5326]      *   |   TAIR10             CDS      &lt;NA&gt;         0
##    16     Chr1 [5439, 5899]      *   |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##    17     Chr1 [5439, 5630]      *   |   TAIR10             CDS      &lt;NA&gt;         0
##    18     Chr1 [5631, 5899]      *   |   TAIR10 three_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##                        ID        Name            Note                          Parent       Index
##               &lt;character&gt; &lt;character&gt; &lt;CharacterList&gt;                 &lt;CharacterList&gt; &lt;character&gt;
##     2         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##     3         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##     4 AT1G01010.1-Protein AT1G01010.1                                                        &lt;NA&gt;
##     5                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##     6                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##   ...                 ...         ...             ...                             ...         ...
##    14                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##    15                &lt;NA&gt;        &lt;NA&gt;                 AT1G01010.1,AT1G01010.1-Protein        &lt;NA&gt;
##    16                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##    17                &lt;NA&gt;        &lt;NA&gt;                 AT1G01010.1,AT1G01010.1-Protein        &lt;NA&gt;
##    18                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##       Derives_from
##        &lt;character&gt;
##     2         &lt;NA&gt;
##     3         &lt;NA&gt;
##     4  AT1G01010.1
##     5         &lt;NA&gt;
##     6         &lt;NA&gt;
##   ...          ...
##    14         &lt;NA&gt;
##    15         &lt;NA&gt;
##    16         &lt;NA&gt;
##    17         &lt;NA&gt;
##    18         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<h2 id="grangeslist-objects">GRangesList Objects</h2>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sp</span> <span class="o">&lt;-</span> <span class="n">split</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span> <span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">gff</span><span class="p">))</span> <span class="c1"># Stores every range in separate component of a GRangesList object
</span><span class="n">split</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span> <span class="n">seqnames</span><span class="p">(</span><span class="n">gff</span><span class="p">))</span> <span class="c1"># Stores ranges of each chromosome in separate component.
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRangesList object of length 7:
## $Chr1 
## GRanges object with 71 ranges and 10 metadata columns:
##       seqnames         ranges strand   |   source            type     score     phase
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;factor&gt;        &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##     2     Chr1   [3631, 5899]      *   |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##     3     Chr1   [3631, 5899]      *   |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##     4     Chr1   [3760, 5630]      *   |   TAIR10         protein      &lt;NA&gt;      &lt;NA&gt;
##     5     Chr1   [3631, 3913]      *   |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##     6     Chr1   [3631, 3759]      *   |   TAIR10  five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##   ...      ...            ...    ... ...      ...             ...       ...       ...
##    68     Chr1 [13335, 13714]      *   |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##    69     Chr1 [12941, 13173]      *   |   TAIR10  five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##    70     Chr1 [11864, 12940]      *   |   TAIR10             CDS      &lt;NA&gt;         0
##    71     Chr1 [11649, 11863]      *   |   TAIR10 three_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##    72     Chr1 [11649, 13173]      *   |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##                        ID        Name            Note                          Parent       Index
##               &lt;character&gt; &lt;character&gt; &lt;CharacterList&gt;                 &lt;CharacterList&gt; &lt;character&gt;
##     2         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##     3         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##     4 AT1G01010.1-Protein AT1G01010.1                                                        &lt;NA&gt;
##     5                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##     6                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##   ...                 ...         ...             ...                             ...         ...
##    68                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##    69                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##    70                &lt;NA&gt;        &lt;NA&gt;                 AT1G01030.1,AT1G01030.1-Protein        &lt;NA&gt;
##    71                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##    72                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##       Derives_from
##        &lt;character&gt;
##     2         &lt;NA&gt;
##     3         &lt;NA&gt;
##     4  AT1G01010.1
##     5         &lt;NA&gt;
##     6         &lt;NA&gt;
##   ...          ...
##    68         &lt;NA&gt;
##    69         &lt;NA&gt;
##    70         &lt;NA&gt;
##    71         &lt;NA&gt;
##    72         &lt;NA&gt;
## 
## ...
## &lt;6 more elements&gt;
## -------
## seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">unlist</span><span class="p">(</span><span class="n">sp</span><span class="p">)</span> <span class="c1"># Returns data as GRanges object
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 442 ranges and 10 metadata columns:
##           seqnames         ranges strand   |   source           type     score     phase
##              &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##       1.2     Chr1   [3631, 5899]      *   |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##       2.3     Chr1   [3631, 5899]      *   |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##       3.4     Chr1   [3760, 5630]      *   |   TAIR10        protein      &lt;NA&gt;      &lt;NA&gt;
##       4.5     Chr1   [3631, 3913]      *   |   TAIR10           exon      &lt;NA&gt;      &lt;NA&gt;
##       5.6     Chr1   [3631, 3759]      *   |   TAIR10 five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##       ...      ...            ...    ... ...      ...            ...       ...       ...
##   438.445     ChrM [11918, 12241]      *   |   TAIR10           gene      &lt;NA&gt;      &lt;NA&gt;
##   439.446     ChrM [11918, 12241]      *   |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##   440.447     ChrM [11918, 12241]      *   |   TAIR10        protein      &lt;NA&gt;      &lt;NA&gt;
##   441.448     ChrM [11918, 12241]      *   |   TAIR10           exon      &lt;NA&gt;      &lt;NA&gt;
##   442.449     ChrM [11918, 12241]      *   |   TAIR10            CDS      &lt;NA&gt;         0
##                            ID        Name                Note                          Parent
##                   &lt;character&gt; &lt;character&gt;     &lt;CharacterList&gt;                 &lt;CharacterList&gt;
##       1.2         AT1G01010.1 AT1G01010.1                                           AT1G01010
##       2.3         AT1G01010.1 AT1G01010.1                                           AT1G01010
##       3.4 AT1G01010.1-Protein AT1G01010.1                                                    
##       4.5                &lt;NA&gt;        &lt;NA&gt;                                         AT1G01010.1
##       5.6                &lt;NA&gt;        &lt;NA&gt;                                         AT1G01010.1
##       ...                 ...         ...                 ...                             ...
##   438.445           ATMG00030   ATMG00030 protein_coding_gene                                
##   439.446         ATMG00030.1 ATMG00030.1                                           ATMG00030
##   440.447 ATMG00030.1-Protein ATMG00030.1                                                    
##   441.448                &lt;NA&gt;        &lt;NA&gt;                                         ATMG00030.1
##   442.449                &lt;NA&gt;        &lt;NA&gt;                     ATMG00030.1,ATMG00030.1-Protein
##                 Index Derives_from
##           &lt;character&gt;  &lt;character&gt;
##       1.2           1         &lt;NA&gt;
##       2.3           1         &lt;NA&gt;
##       3.4        &lt;NA&gt;  AT1G01010.1
##       4.5        &lt;NA&gt;         &lt;NA&gt;
##       5.6        &lt;NA&gt;         &lt;NA&gt;
##       ...         ...          ...
##   438.445        &lt;NA&gt;         &lt;NA&gt;
##   439.446           1         &lt;NA&gt;
##   440.447        &lt;NA&gt;  ATMG00030.1
##   441.448        &lt;NA&gt;         &lt;NA&gt;
##   442.449        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sp</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span> <span class="s2">"type"</span><span class="p">]</span> <span class="c1"># Subsetting of GRangesList objects is similar to GRanges objects.
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRangesList object of length 4:
## $1 
## GRanges object with 1 range and 1 metadata column:
##     seqnames       ranges strand |     type
##        &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   2     Chr1 [3631, 5899]      * |     mRNA
## 
## $2 
## GRanges object with 1 range and 1 metadata column:
##     seqnames       ranges strand | type
##   3     Chr1 [3631, 5899]      * | mRNA
## 
## $3 
## GRanges object with 1 range and 1 metadata column:
##     seqnames       ranges strand |    type
##   4     Chr1 [3760, 5630]      * | protein
## 
## ...
## &lt;1 more element&gt;
## -------
## seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">lapply</span><span class="p">(</span><span class="n">sp</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span> <span class="n">length</span><span class="p">)</span> <span class="c1"># Looping over GRangesList objects similar to lists
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## $`1`
## [1] 1
## 
## $`2`
## [1] 1
## 
## $`3`
## [1] 1
## 
## $`4`
## [1] 1
</code></pre>
</div>

<h1 id="transcript-ranges">Transcript Ranges</h1>

<p>Storing annotation ranges in <code class="highlighter-rouge">TranscriptDb</code> databases makes many operations more robust and convenient.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">GenomicFeatures</span><span class="p">)</span>
<span class="n">download.file</span><span class="p">(</span><span class="s2">"http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/gff3.gff"</span><span class="p">,</span> <span class="s2">"data/gff3.gff"</span><span class="p">)</span>
<span class="n">txdb</span> <span class="o">&lt;-</span> <span class="n">makeTxDbFromGFF</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"data/gff3.gff"</span><span class="p">,</span> <span class="n">format</span><span class="o">=</span><span class="s2">"gff"</span><span class="p">,</span> <span class="n">dataSource</span><span class="o">=</span><span class="s2">"TAIR"</span><span class="p">,</span> <span class="n">organism</span><span class="o">=</span><span class="s2">"Arabidopsis thaliana"</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning in .extract_exons_from_GRanges(cds_IDX, gr, ID, Name, Parent, feature = "cds", : The following orphan CDS were dropped (showing only the 6 first):
##   seqid start  end strand   ID              Parent Name
## 1  Chr1  3760 3913      + &lt;NA&gt; AT1G01010.1-Protein &lt;NA&gt;
## 2  Chr1  3996 4276      + &lt;NA&gt; AT1G01010.1-Protein &lt;NA&gt;
## 3  Chr1  4486 4605      + &lt;NA&gt; AT1G01010.1-Protein &lt;NA&gt;
## 4  Chr1  4706 5095      + &lt;NA&gt; AT1G01010.1-Protein &lt;NA&gt;
## 5  Chr1  5174 5326      + &lt;NA&gt; AT1G01010.1-Protein &lt;NA&gt;
## 6  Chr1  5439 5630      + &lt;NA&gt; AT1G01010.1-Protein &lt;NA&gt;
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">saveDb</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s2">"./data/TAIR10.sqlite"</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: TAIR
## # Organism: Arabidopsis thaliana
## # Taxonomy ID: 3702
## # miRBase build ID: NA
## # Genome: NA
## # transcript_nrow: 28
## # exon_nrow: 113
## # cds_nrow: 99
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2016-04-21 12:28:05 -0700 (Thu, 21 Apr 2016)
## # GenomicFeatures version at creation time: 1.22.6
## # RSQLite version at creation time: 1.0.0
## # DBSCHEMAVERSION: 1.1
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">txdb</span> <span class="o">&lt;-</span> <span class="n">loadDb</span><span class="p">(</span><span class="s2">"./data/TAIR10.sqlite"</span><span class="p">)</span>
<span class="n">transcripts</span><span class="p">(</span><span class="n">txdb</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 28 ranges and 2 metadata columns:
##        seqnames         ranges strand   |     tx_id     tx_name
##           &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;   | &lt;integer&gt; &lt;character&gt;
##    [1]     Chr1 [ 3631,  5899]      +   |         1 AT1G01010.1
##    [2]     Chr1 [ 5928,  8737]      -   |         2 AT1G01020.1
##    [3]     Chr1 [ 6790,  8737]      -   |         3 AT1G01020.2
##    [4]     Chr1 [11649, 13714]      -   |         4 AT1G01030.1
##    [5]     Chr2 [ 1025,  2810]      +   |         5 AT2G01008.1
##    ...      ...            ...    ... ...       ...         ...
##   [24]     ChrC [  383,  1444]      -   |        24 ATCG00020.1
##   [25]     ChrC [ 1717,  4347]      -   |        25 ATCG00030.1
##   [26]     ChrM [11918, 12241]      +   |        26 ATMG00030.1
##   [27]     ChrM [  273,   734]      -   |        27 ATMG00010.1
##   [28]     ChrM [ 8848, 11415]      -   |        28 ATMG00020.1
##   -------
##   seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">transcriptsBy</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">"gene"</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRangesList object of length 22:
## $AT1G01010 
## GRanges object with 1 range and 2 metadata columns:
##       seqnames       ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     Chr1 [3631, 5899]      + |         1 AT1G01010.1
## 
## $AT1G01020 
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames       ranges strand | tx_id     tx_name
##   [1]     Chr1 [5928, 8737]      - |     2 AT1G01020.1
##   [2]     Chr1 [6790, 8737]      - |     3 AT1G01020.2
## 
## $AT1G01030 
## GRanges object with 1 range and 2 metadata columns:
##       seqnames         ranges strand | tx_id     tx_name
##   [1]     Chr1 [11649, 13714]      - |     4 AT1G01030.1
## 
## ...
## &lt;19 more elements&gt;
## -------
## seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">exonsBy</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span> <span class="n">by</span> <span class="o">=</span> <span class="s2">"gene"</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRangesList object of length 22:
## $AT1G01010 
## GRanges object with 6 ranges and 2 metadata columns:
##       seqnames       ranges strand |   exon_id   exon_name
##          &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     Chr1 [3631, 3913]      + |         1        &lt;NA&gt;
##   [2]     Chr1 [3996, 4276]      + |         2        &lt;NA&gt;
##   [3]     Chr1 [4486, 4605]      + |         3        &lt;NA&gt;
##   [4]     Chr1 [4706, 5095]      + |         4        &lt;NA&gt;
##   [5]     Chr1 [5174, 5326]      + |         5        &lt;NA&gt;
##   [6]     Chr1 [5439, 5899]      + |         6        &lt;NA&gt;
## 
## $AT1G01020 
## GRanges object with 12 ranges and 2 metadata columns:
##        seqnames       ranges strand   | exon_id exon_name
##    [1]     Chr1 [5928, 6263]      -   |       7      &lt;NA&gt;
##    [2]     Chr1 [6437, 7069]      -   |       8      &lt;NA&gt;
##    [3]     Chr1 [6790, 7069]      -   |       9      &lt;NA&gt;
##    [4]     Chr1 [7157, 7232]      -   |      10      &lt;NA&gt;
##    [5]     Chr1 [7157, 7450]      -   |      11      &lt;NA&gt;
##    ...      ...          ...    ... ...     ...       ...
##    [8]     Chr1 [7762, 7835]      -   |      14      &lt;NA&gt;
##    [9]     Chr1 [7942, 7987]      -   |      15      &lt;NA&gt;
##   [10]     Chr1 [8236, 8325]      -   |      16      &lt;NA&gt;
##   [11]     Chr1 [8417, 8464]      -   |      17      &lt;NA&gt;
##   [12]     Chr1 [8571, 8737]      -   |      18      &lt;NA&gt;
## 
## $AT1G01030 
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames         ranges strand | exon_id exon_name
##   [1]     Chr1 [11649, 13173]      - |      19      &lt;NA&gt;
##   [2]     Chr1 [13335, 13714]      - |      20      &lt;NA&gt;
## 
## ...
## &lt;19 more elements&gt;
## -------
## seqinfo: 7 sequences (2 circular) from an unspecified genome; no seqlengths
</code></pre>
</div>

<h2 id="txdb-from-biomart"><code class="highlighter-rouge">txdb</code> from BioMart</h2>

<p>Alternative sources for creating <code class="highlighter-rouge">txdb</code> databases are BioMart, Bioc annotation packages, UCSC, etc. The following shows how to create a <code class="highlighter-rouge">txdb</code> from BioMart.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">GenomicFeatures</span><span class="p">);</span> <span class="n">library</span><span class="p">(</span><span class="s2">"biomaRt"</span><span class="p">)</span>
<span class="n">txdb</span> <span class="o">&lt;-</span> <span class="n">makeTxDbFromBiomart</span><span class="p">(</span><span class="n">biomart</span> <span class="o">=</span> <span class="s2">"plants_mart"</span><span class="p">,</span> <span class="n">dataset</span> <span class="o">=</span> <span class="s2">"athaliana_eg_gene"</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s2">"plants.ensembl.org"</span><span class="p">)</span>
</code></pre>
</div>

<p>The following steps are useful to find out what is availble in BioMart.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">listMarts</span><span class="p">()</span> <span class="c1"># Lists BioMart databases
</span><span class="n">listMarts</span><span class="p">(</span><span class="n">host</span><span class="o">=</span><span class="s2">"plants.ensembl.org"</span><span class="p">)</span>
<span class="n">mymart</span> <span class="o">&lt;-</span> <span class="n">useMart</span><span class="p">(</span><span class="s2">"plants_mart"</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s2">"plants.ensembl.org"</span><span class="p">)</span> <span class="c1"># Select one, here plants_mart_25
</span><span class="n">listDatasets</span><span class="p">(</span><span class="n">mymart</span><span class="p">)</span> <span class="c1"># List datasets available in the selected BioMart database
</span><span class="n">mymart</span> <span class="o">&lt;-</span> <span class="n">useMart</span><span class="p">(</span><span class="s2">"plants_mart"</span><span class="p">,</span> <span class="n">dataset</span><span class="o">=</span><span class="s2">"athaliana_eg_gene"</span><span class="p">,</span> <span class="n">host</span><span class="o">=</span><span class="s2">"plants.ensembl.org"</span><span class="p">)</span>
<span class="n">listAttributes</span><span class="p">(</span><span class="n">mymart</span><span class="p">)</span> <span class="c1"># List available features 
</span><span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">"ensembl_gene_id"</span><span class="p">,</span> <span class="s2">"description"</span><span class="p">),</span> <span class="n">mart</span><span class="o">=</span><span class="n">mymart</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span>
</code></pre>
</div>

<h2 id="efficient-sequence-parsing-with-getseq">Efficient Sequence Parsing with <code class="highlighter-rouge">getSeq</code></h2>

<p>The following parses all annotation ranges provided by <code class="highlighter-rouge">GRanges</code> object (e.g. <code class="highlighter-rouge">gff</code>) from a genome sequence stored in a local file.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span> <span class="o">&lt;-</span> <span class="n">gff</span><span class="p">[</span><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="o">$</span><span class="n">type</span> <span class="o">!=</span> <span class="s2">"chromosome"</span><span class="p">]</span> <span class="c1"># Remove chromosome ranges
</span><span class="n">rand</span> <span class="o">&lt;-</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">sapply</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">as.character</span><span class="p">(</span><span class="n">seqnames</span><span class="p">(</span><span class="n">gff</span><span class="p">))),</span> <span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">paste</span><span class="p">(</span><span class="n">sample</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="s2">"A"</span><span class="p">,</span><span class="s2">"T"</span><span class="p">,</span><span class="s2">"G"</span><span class="p">,</span><span class="s2">"C"</span><span class="p">),</span> <span class="m">200000</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="n">T</span><span class="p">),</span> <span class="n">collapse</span><span class="o">=</span><span class="s2">""</span><span class="p">)))</span>
<span class="n">writeXStringSet</span><span class="p">(</span><span class="n">DNAStringSet</span><span class="p">(</span><span class="n">rand</span><span class="p">),</span> <span class="s2">"./data/test"</span><span class="p">)</span>
<span class="n">getSeq</span><span class="p">(</span><span class="n">FaFile</span><span class="p">(</span><span class="s2">"./data/test"</span><span class="p">),</span> <span class="n">gff</span><span class="p">)</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   A DNAStringSet instance of length 442
##       width seq                                                                 names               
##   [1]  2269 TAATCCCGCCAGGGGTACCTTACAATGAGACG...CGTCGCAGCTAAACCTAGTATGTTCTATATAC Chr1
##   [2]  2269 TAATCCCGCCAGGGGTACCTTACAATGAGACG...CGTCGCAGCTAAACCTAGTATGTTCTATATAC Chr1
##   [3]  1871 TTCAACGTGAGTCCCCAAACTCCGTCGTCAAG...TCTACGAAATGTGTGCGGAGGACCATATCCGT Chr1
##   [4]   283 TAATCCCGCCAGGGGTACCTTACAATGAGACG...ATGCAGTAATTTCCGCTGTCTCTAGCCTAAAT Chr1
##   [5]   129 TAATCCCGCCAGGGGTACCTTACAATGAGACG...ACTGCTCATACCCGGCACTTTATCCCCCGCAG Chr1
##   ...   ... ...
## [438]   324 TCCCCTCCCAGCCTCCTAGAGGTGACGAGGAT...GACGGGGGGAAATGCTGTAGTCTCAAATGACC ChrM
## [439]   324 TCCCCTCCCAGCCTCCTAGAGGTGACGAGGAT...GACGGGGGGAAATGCTGTAGTCTCAAATGACC ChrM
## [440]   324 TCCCCTCCCAGCCTCCTAGAGGTGACGAGGAT...GACGGGGGGAAATGCTGTAGTCTCAAATGACC ChrM
## [441]   324 TCCCCTCCCAGCCTCCTAGAGGTGACGAGGAT...GACGGGGGGAAATGCTGTAGTCTCAAATGACC ChrM
## [442]   324 TCCCCTCCCAGCCTCCTAGAGGTGACGAGGAT...GACGGGGGGAAATGCTGTAGTCTCAAATGACC ChrM
</code></pre>
</div>

<h1 id="homework-6">Homework 6</h1>

<h2 id="hw6a---demultiplexing">HW6a - Demultiplexing</h2>

<p>Write a demultiplexing function that accepts any number of
barcodes and splits a FASTQ file into as many subfiles as there are barcodes.
At the same time the function should remove low quality tails from the reads.
The following function accomplishes the first step. Expand this function so
that it performs the second step as well.</p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">demultiplex</span> <span class="o">&lt;-</span> <span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">barcode</span><span class="p">,</span> <span class="n">nreads</span><span class="p">)</span> <span class="p">{</span>
	<span class="n">f</span> <span class="o">&lt;-</span> <span class="n">FastqStreamer</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">nreads</span><span class="p">)</span> 
	<span class="k">while</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">fq</span> <span class="o">&lt;-</span> <span class="n">yield</span><span class="p">(</span><span class="n">f</span><span class="p">)))</span> <span class="p">{</span>
		<span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="k">in</span> <span class="n">barcode</span><span class="p">)</span> <span class="p">{</span>
			<span class="n">pattern</span> <span class="o">&lt;-</span> <span class="n">paste</span><span class="p">(</span><span class="s2">"^"</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span>
			<span class="n">fqsub</span> <span class="o">&lt;-</span> <span class="n">fq</span><span class="p">[</span><span class="n">grepl</span><span class="p">(</span><span class="n">pattern</span><span class="p">,</span> <span class="n">sread</span><span class="p">(</span><span class="n">fq</span><span class="p">))]</span> 
			<span class="k">if</span><span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">fqsub</span><span class="p">)</span> <span class="o">&gt;</span> <span class="m">0</span><span class="p">)</span> <span class="p">{</span>
				<span class="n">writeFastq</span><span class="p">(</span><span class="n">fqsub</span><span class="p">,</span> <span class="n">paste</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">i</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">"_"</span><span class="p">),</span> <span class="n">mode</span><span class="o">=</span><span class="s2">"a"</span><span class="p">,</span> <span class="n">compress</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
			<span class="p">}</span>
		<span class="p">}</span>
	<span class="p">}</span>
	<span class="n">close</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="p">}</span>
<span class="n">demultiplex</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">fastq</span><span class="p">[</span><span class="m">1</span><span class="p">],</span> <span class="n">barcode</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">"TT"</span><span class="p">,</span> <span class="s2">"AA"</span><span class="p">,</span> <span class="s2">"GG"</span><span class="p">),</span> <span class="n">nreads</span><span class="o">=</span><span class="m">50</span><span class="p">)</span>
</code></pre>
</div>

<h2 id="hw6b---sequence-parsing">HW6b - Sequence Parsing</h2>

<ul>
  <li>Download <code class="highlighter-rouge">GFF</code> from <em>Halobacterium sp</em>  <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.gff">here</a></li>
  <li>Download genome sequence from -Halobacterium sp_ <a href="ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.fna">here</a></li>
  <li><strong>Task 1</strong> Extract gene ranges, parse their sequences from genome and translate them into proteins</li>
  <li><strong>Task 2</strong> Reduce overlapping genes and parse their sequences from genome</li>
  <li><strong>Task 3</strong> Generate intergenic ranges and parse their sequences from genome</li>
</ul>

<p><strong>Useful commands</strong></p>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">download.file</span><span class="p">(</span><span class="s2">"ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.gff"</span><span class="p">,</span> <span class="s2">"data/AE004437.gff"</span><span class="p">)</span>
<span class="n">download.file</span><span class="p">(</span><span class="s2">"ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Bacteria/Halobacterium_sp_uid217/AE004437.fna"</span><span class="p">,</span> <span class="s2">"data/AE004437.fna"</span><span class="p">)</span>
<span class="n">chr</span> <span class="o">&lt;-</span> <span class="n">readDNAStringSet</span><span class="p">(</span><span class="s2">"data/AE004437.fna"</span><span class="p">)</span>
<span class="n">gff</span> <span class="o">&lt;-</span> <span class="n">import</span><span class="p">(</span><span class="s2">"data/AE004437.gff"</span><span class="p">)</span>
<span class="n">gffgene</span> <span class="o">&lt;-</span> <span class="n">gff</span><span class="p">[</span><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)[,</span><span class="s2">"type"</span><span class="p">]</span><span class="o">==</span><span class="s2">"gene"</span><span class="p">]</span>
<span class="n">gene</span> <span class="o">&lt;-</span> <span class="n">DNAStringSet</span><span class="p">(</span><span class="n">Views</span><span class="p">(</span><span class="n">chr</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span> <span class="n">IRanges</span><span class="p">(</span><span class="n">start</span><span class="p">(</span><span class="n">gffgene</span><span class="p">),</span> <span class="n">end</span><span class="p">(</span><span class="n">gffgene</span><span class="p">))))</span>
<span class="n">names</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">values</span><span class="p">(</span><span class="n">gffgene</span><span class="p">)[,</span><span class="s2">"locus_tag"</span><span class="p">]</span>
<span class="n">pos</span> <span class="o">&lt;-</span> <span class="n">values</span><span class="p">(</span><span class="n">gffgene</span><span class="p">[</span><span class="n">strand</span><span class="p">(</span><span class="n">gffgene</span><span class="p">)</span> <span class="o">==</span> <span class="s2">"+"</span><span class="p">])[,</span><span class="s2">"locus_tag"</span><span class="p">]</span>
<span class="n">p1</span> <span class="o">&lt;-</span> <span class="n">translate</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="n">names</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">%</span><span class="k">in</span><span class="o">%</span> <span class="n">pos</span><span class="p">])</span>
<span class="n">names</span><span class="p">(</span><span class="n">p1</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="n">names</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">%</span><span class="k">in</span><span class="o">%</span> <span class="n">pos</span><span class="p">])</span>
<span class="n">neg</span> <span class="o">&lt;-</span> <span class="n">values</span><span class="p">(</span><span class="n">gffgene</span><span class="p">[</span><span class="n">strand</span><span class="p">(</span><span class="n">gffgene</span><span class="p">)</span> <span class="o">==</span> <span class="s2">"-"</span><span class="p">])[,</span><span class="s2">"locus_tag"</span><span class="p">]</span>
<span class="n">p2</span> <span class="o">&lt;-</span> <span class="n">translate</span><span class="p">(</span><span class="n">reverseComplement</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="n">names</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">%</span><span class="k">in</span><span class="o">%</span> <span class="n">neg</span><span class="p">]))</span>
<span class="n">names</span><span class="p">(</span><span class="n">p2</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">names</span><span class="p">(</span><span class="n">gene</span><span class="p">[</span><span class="n">names</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span> <span class="o">%</span><span class="k">in</span><span class="o">%</span> <span class="n">neg</span><span class="p">])</span>
<span class="n">writeXStringSet</span><span class="p">(</span><span class="n">c</span><span class="p">(</span><span class="n">p1</span><span class="p">,</span> <span class="n">p2</span><span class="p">),</span> <span class="s2">"./data/mypep.fasta"</span><span class="p">)</span>
</code></pre>
</div>

<h2 id="homework-submission">Homework submission</h2>
<p>Submit the homework results in one well structured and annotated R script to the instructor. The script should include instructions on how to use the functions.</p>

<h2 id="due-date">Due date</h2>

<p>This homework is due on Thu, April 28th at 6:00 PM.</p>

<h2 id="homework-solutions">Homework Solutions</h2>

<p>See <a href="https://drive.google.com/file/d/0B-lLYVUOliJFUmNlX2twQjBjaW8/view?usp=sharing">here</a></p>

<h1 id="session-info">Session Info</h1>

<div class="highlighter-rouge"><pre class="highlight"><code><span class="n">sessionInfo</span><span class="p">()</span>
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## R version 3.2.4 Revised (2016-03-16 r70336)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.4 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] grid      stats4    parallel  stats     graphics  utils     datasets  grDevices methods  
## [10] base     
## 
## other attached packages:
##  [1] GenomicFeatures_1.22.6     AnnotationDbi_1.32.0       rtracklayer_1.30.1        
##  [4] systemPipeR_1.4.8          RSQLite_1.0.0              DBI_0.3.1                 
##  [7] ShortRead_1.28.0           GenomicAlignments_1.6.1    SummarizedExperiment_1.0.1
## [10] Biobase_2.30.0             Rsamtools_1.22.0           GenomicRanges_1.22.1      
## [13] GenomeInfoDb_1.6.1         BiocParallel_1.4.0         seqLogo_1.36.0            
## [16] Biostrings_2.38.2          XVector_0.10.0             IRanges_2.4.4             
## [19] S4Vectors_0.8.3            BiocGenerics_0.16.1        ggplot2_2.0.0             
## [22] limma_3.26.3               BiocStyle_1.8.0           
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.3            lattice_0.20-33        GO.db_3.2.2            digest_0.6.9          
##  [5] plyr_1.8.3             futile.options_1.0.0   BatchJobs_1.6          evaluate_0.8          
##  [9] zlibbioc_1.16.0        annotate_1.48.0        Matrix_1.2-3           checkmate_1.6.3       
## [13] rmarkdown_0.9.2        labeling_0.3           GOstats_2.36.0         splines_3.2.4         
## [17] stringr_1.0.0          pheatmap_1.0.7         RCurl_1.95-4.7         biomaRt_2.26.1        
## [21] munsell_0.4.2          sendmailR_1.2-1        base64enc_0.1-3        BBmisc_1.9            
## [25] htmltools_0.3          fail_1.3               edgeR_3.12.0           codetools_0.2-14      
## [29] XML_3.98-1.3           AnnotationForge_1.12.0 bitops_1.0-6           RBGL_1.46.0           
## [33] xtable_1.8-0           GSEABase_1.32.0        gtable_0.1.2           magrittr_1.5          
## [37] formatR_1.2.1          scales_0.3.0           graph_1.48.0           stringi_1.0-1         
## [41] hwriter_1.3.2          genefilter_1.52.0      latticeExtra_0.6-26    futile.logger_1.4.1   
## [45] brew_1.0-6             rjson_0.2.15           lambda.r_1.1.7         RColorBrewer_1.1-2    
## [49] tools_3.2.4            Category_2.36.0        survival_2.38-3        yaml_2.1.13           
## [53] colorspace_1.2-6       knitr_1.12
</code></pre>
</div>

<h1 id="references">References</h1>



<div class="tags">
    
</div>



</div>



<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2016 UC Riverside. All rights reserved. <br />
 Site last generated: Aug 11, 2016 <br />
<p><img src="../common_images/company_logo.png/></p>
                </div>
            </div>
</footer>

            </div><!-- /.row -->

    </div>    <!-- /.container -->

</body>


</html>

