<!DOCTYPE html>
  <head>

 <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>VAR-Seq Workflow Template  | GEN242</title>
<link rel="stylesheet" type="text/css" href="../css/syntax.css">
<link rel="stylesheet" type="text/css" href="../css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="../css/bootstrap.min.css">-->
<link rel="stylesheet" type="text/css" href="../css/modern-business.css">
<link rel="stylesheet" type="text/css" href="../css/lavish-bootstrap.css">
<link rel="stylesheet" type="text/css" href="../css/customstyles.css">
<link rel="stylesheet" type="text/css" href="../css/theme-blue.css">
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="../js/jquery.navgoco.min.js"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="../js/toc.js"></script>
<script src="../js/customscripts.js"></script>
<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<link rel="shortcut icon" href="../common_images/favicon.ico" type="image/x-icon">
<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->







 


<script>
  $(function () {
      $('[data-toggle="tooltip"]').tooltip()
  })
</script>



  </head>

<body>

   <!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>

            <a class="fa fa-home fa-lg navbar-brand" href="home.html">&nbsp;<span class="projectTitle"> Data Analysis in Genome Biology</span></a>

        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- entries without drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                



















                
                
                
                
                <li><a href="https://piazza.com/class/ilbslqb3j2ovy" target="_blank">Piazza</a></li>
                
                
                
                
                
                <li><a href="https://github.com/tgirke/GEN242" target="_blank">Github Repo</a></li>
                
                
                
                


                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->

                <li class="dropdown">
                    
                    
                    
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://jekyllrb.com/" target="_blank">Jekyll</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                        
                        

                    </ul>
                </li>
                
                


                <!-- special insertion -->

               
                
                
                <!-- Send feedback function -->
<script>
function SendLinkByMail(href) {
var subject= "GEN242 feedback";
var body = "I have some feedback about the VAR-Seq Workflow Template page: ";
body += window.location.href;
body += "";
var uri = "mailto:?subject=";
uri += encodeURIComponent(subject);
uri += "&body=";
uri += encodeURIComponent(body);
window.location.href = uri;
}
</script>

<li><a href="javascript:(function()%7BSendLinkByMail()%3B%7D)()%3B" target="_blank"><i class="fa fa-envelope-o"></i> Feedback</a></li>


                <!--uncomment this block if you want simple search instead of algolia-->
                <li>
                     <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="../js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                        SimpleJekyllSearch.init({
                            searchInput: document.getElementById('search-input'),
                            resultsContainer: document.getElementById('results-container'),
                            dataSource: '../search.json',
                            searchResultTemplate: '<li><a href="{url}" title="VAR-Seq Workflow Template">{title}</a></li>',
                        noResultsText: 'No results found.',
                                limit: 10,
                                fuzzy: true,
                        })
                    </script>
                     <!--end search-->
                </li>

            
        </div>
        <!-- /.container -->
</nav>


    <!-- Page Content -->
    <div class="container">
        <div class="col-lg-12">&nbsp;</div>


<!-- Content Row -->
<div class="row">
    <!-- Sidebar Column -->
    <div class="col-md-3">

        <script>

            $(document).ready(function() {
                // Initialize navgoco with default options
                $("#mysidebar").navgoco({
                    caretHtml: '',
                    accordion: true,
                    openClass: 'active', // open
                    save: false, // leave false or nav highlighting doesn't work right
                    cookie: {
                        name: 'navgoco',
                        expires: false,
                        path: '/'
                    },
                    slide: {
                        duration: 400,
                        easing: 'swing'
                    }
                });

                $("#collapseAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', false);
                });

                $("#expandAll").click(function(e) {
                    e.preventDefault();
                    $("#mysidebar").navgoco('toggle', true);
                });

            });

        </script>


        



















        <ul id="mysidebar" class="nav">

            <span class="siteTagline">GEN242</span>
            <span class="versionTagline">Spring 2016</span>

            
            
            
                
            
            <li><a href="#">Home</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/home.html">Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_syllabus.html">Syllabus</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Internal</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_internal.html">Internal Resources</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Schedule</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_schedule.html">Course Schedule</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Lecture Slides</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_01.html">Course Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_02.html">Molbiol for Bioinformatics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_03.html">Databases and Software</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_04.html">DNA Sequencing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_05.html">Introduction to R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_06.html">Sequence Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_07.html">Programming in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_08.html">Multiple Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_09.html">Short Read Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_10.html">NGS Analysis Basics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_11.html">Gene Expression Analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_12.html">NGS Workflows</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_13.html">ChIP-Seq Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_14.html">VAR-Seq Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_15.html">Sequence Assembly</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_16.html">Gene Annotation and Ontologies</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_slides_17.html">Cluster Analysis</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Homework</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_01.html">HW1 - Online Tools</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_02.html">HW2 - Biocluster/Linux</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_03.html">HW3 - Introduction to R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_homework_04.html">HW4 - Pairwise Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_09.html">HW5 - Programming in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_08.html">HW6 - Sequence Analysis</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">GitHub</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_tutorial_01.html">GitHub</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Biocluster & Linux</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_tutorial_02.html">Basic Tutorial</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Introduction</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_03.html">3. Installation of R and Add-on Packages</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_04.html">4. Getting Around</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_06.html">6. Data Types</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_07.html">7. Data objects</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_08.html">8. Important Utilities</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_12.html">12. SQLite Databases</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_13.html">13. Graphics in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_14.html">14. Analysis Routine</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_15.html">15. R Markdown</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_16.html">16. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_17.html">17. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Introduction to R</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_03.html">3. Installation of R and Add-on Packages</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_04.html">4. Getting Around</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_06.html">6. Data Types</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_07.html">7. Data objects</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_08.html">8. Important Utilities</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_12.html">12. SQLite Databases</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_13.html">13. Graphics in R</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_14.html">14. Analysis Routine</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_15.html">15. R Markdown</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_16.html">16. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rbasics_17.html">17. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Programming in R</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_02.html">2. Control Structures</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_03.html">3. Loops</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_04.html">4. Functions</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_05.html">5. Useful Utilities</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_06.html">6. Running R Scripts</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_07.html">7. Building R Packages</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_08.html">8. Programming Exercises</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_09.html">9. Homework 5</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_10.html">10. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Programming_in_R_11.html">11. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">NGS Analysis Basics</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_01.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_02.html">2. Package Requirements</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_03.html">3. Strings in R Base</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_04.html">4. Sequences in Bioconductor</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_05.html">5. NGS Sequences</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_06.html">6. Range Operations</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_07.html">7. Transcript Ranges</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_08.html">8. Homework 6</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_09.html">9. Session Info</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rsequences_10.html">10. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">NGS Workflow Overview</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_1.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_2.html">2. Getting Started</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_5.html">5. Version information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeR_6.html">6. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">RNA-Seq Workflow</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_02.html">2. Sample definitions and environment settings</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_05.html">5. Read quantification per annotation range</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_06.html">6. Analysis of differentially expressed genes with edgeR</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis of DEGs</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeRNAseq_11.html">11. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">ChIP-Seq Workflow</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_02.html">2. Load workflow environment</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeChIPseq_14.html">14. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">VAR-Seq Workflow</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_02.html">2. Load workflow environment</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_systemPipeVARseq_14.html">14. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Graphics and Visualization</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_1.html">1. Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_2.html">2. Base Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_3.html">3. Grid Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_4.html">4. lattice Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_5.html">5. ggplot2 Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_6.html">6. Specialty Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_7.html">7. Genome Graphics</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rgraphics_8.html">8. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Cluster Analysis</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_1.html">1. Introduction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_2.html">2. Data Preprocessing</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_3.html">3. Clustering Algorithms</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_4.html">4. Clustering Exercises</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_5.html">5. Version Information</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_Rclustering_6.html">6. References</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Paper Presentations</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_paper_presentations.html">Paper Presentations</a></li>
                    

                    

                    
                    
                </ul>
                
                
            
            <li><a href="#">Course Projects</a>
                <ul>
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_00.html">Overview</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_01.html">ChIP-Seq1 - Motif Prediction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_02.html">ChIP-Seq2 - Peak Callers</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_03.html">RNA-Seq1 - RNA-Seq Aligners</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_04.html">RNA-Seq2 - DEG Methods</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_05.html">VAR-Seq1 - Variant Callers</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_06.html">VAR-Seq2 - Functional Prediction</a></li>
                    

                    

                    
                    
                    
                    
                    <li><a href="../mydoc/mydoc_project_07.html">Project Data</a></li>
                    

                    

                    
                    
                </ul>
                
                
                


                <!-- if you aren't using the accordion, uncomment this block:

                     <p class="external">
                         <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
                     </p>
                 -->

    </div>
    <!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
    <script>$("li.active").parents('li').toggleClass("active");</script>


            <!-- Content Column -->
            <div class="col-md-9">
                
                <div class="post-header">
   <h1 class="post-title-main">VAR-Seq Workflow Template</h1>
</div>

<div class="post-content">

   

<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>


<div id="toc"></div>


<!--
    
    <a target="_blank" href="https://github.com/tgirke/manuals/tree/master/mydoc/" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
    
-->
<!--
%% \VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{VAR-Seq Workflow Template}
%% \VignetteDepends{rjson, ggplot2, limma, edgeR, DESeq2, GOstats, GO.db, annotate, pheatmap}
%% \VignetteKeywords{compute cluster, pipeline, reports}
%% \VignettePackage{systemPipeR}
-->

<!---
- Conversion from Rnw/tex to Rmd
    - Build vignette from Rnw 
    - Then run: pandoc -s systemPipeVARseq.tex -o systemPipeVARseq.text
    - The fix things, usually a lot...

- Compile from command-line
echo "rmarkdown::render('systemPipeVARseq.Rmd', clean=F)" | R -slave; R CMD Stangle systemPipeVARseq.Rmd; Rscript ../md2jekyll.R systemPipeVARseq.knit.md 14

- Commit to github
git commit -am "some edits"; git push -u origin master

- To customize font size and other style features, add this line to output section in preamble:  
    css: style.css
-->

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script>

<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script>

<p>Note: the most recent version of this tutorial can be found <a href="https://htmlpreview.github.io/?https://github.com/tgirke/GEN242/blob/master/vignettes/13_VARseqWorkflow/systemPipeVARseq.html">here</a> and a short overview slide show <a href="https://docs.google.com/presentation/d/1cTMVFvvymIWBG2hNq0GoasSDDyoGfjSiZZny9NHpK0o/edit?usp=sharing">here</a>.</p>

<h1 id="introduction">Introduction</h1>

<p>Users want to provide here background information about the design of their VAR-Seq project.</p>

<h2 id="background-and-objectives">Background and objectives</h2>

<p>This report describes the analysis of a VAR-Seq project studying the
genetic differences among several strains  from <em>organism</em> .</p>

<h2 id="experimental-design">Experimental design</h2>

<p>Typically, users want to specify here all information relevant for the
analysis of their NGS study. This includes detailed descriptions of
FASTQ files, experimental design, reference genome, gene annotations,
etc.</p>

<h1 id="load-workflow-environment">Load workflow environment</h1>

<p>Now open the R markdown script <code>systemPipeVARseq.Rmd</code>in your R IDE (<em>e.g.</em> vim-r or RStudio) and 
run the workflow as outlined below.</p>

<h2 id="load-packages-and-sample-data">Load packages and sample data</h2>

<p>The <code>systemPipeR</code> package needs to be loaded to perform the analysis steps shown in
this report [@Girke2014-oy].</p>

<p><code>r
library(systemPipeR)
</code></p>

<p>Load workflow environment with sample data into your current working
directory. The sample data are described
<a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/systemPipeR/inst/doc/systemPipeR.html#load-sample-data-and-workflow-templates">here</a>.</p>

<p><code>r
library(systemPipeRdata)
genWorkenvir(workflow="varseq")
setwd("varseq")
download.file("https://raw.githubusercontent.com/tgirke/GEN242/master/vignettes/13_VARseqWorkflow/systemPipeVARseq.Rmd", "systemPipeVARseq.Rmd")
</code></p>

<p>In the workflow environments generated by <code>genWorkenvir</code> all data inputs are stored in
a <code>data/</code> directory and all analysis results will be written to a separate
<code>results/</code> directory, while the <code>systemPipeVARseq.Rmd</code> script and the <code>targets</code> file are expected to be located in
the parent directory. The R session is expected to run from this parent
directory. Additional parameter files are stored under <code>param/</code>.</p>

<p>To work with real data, users want to organize their own data similarly
and substitute all test data for their own data. To rerun an established
workflow on new data, the initial <code>targets</code> file along with the corresponding
FASTQ files are usually the only inputs the user needs to provide.</p>

<p>If applicable users can load custom functions not provided by <code>systemPipeR</code>. Skip
this step if this is not the case.</p>

<p><code>r
source("systemPipeChIPseq_Fct.R")
</code></p>

<h2 id="experiment-definition-provided-by-targets-file">Experiment definition provided by <code>targets</code> file</h2>

<p>The <code>targets</code> file defines all FASTQ files and sample comparisons of the analysis workflow.</p>

<p><code>r
targetspath &lt;- system.file("extdata", "targetsPE.txt", package="systemPipeR")
targets &lt;- read.delim(targetspath, comment.char = "#")
targets[,-c(5,6)]
</code></p>

<p><code>
##                   FileName1                FileName2 SampleName Factor        Date
## 1  ./data/SRR446027_1.fastq ./data/SRR446027_2.fastq        M1A     M1 23-Mar-2012
## 2  ./data/SRR446028_1.fastq ./data/SRR446028_2.fastq        M1B     M1 23-Mar-2012
## 3  ./data/SRR446029_1.fastq ./data/SRR446029_2.fastq        A1A     A1 23-Mar-2012
## 4  ./data/SRR446030_1.fastq ./data/SRR446030_2.fastq        A1B     A1 23-Mar-2012
## 5  ./data/SRR446031_1.fastq ./data/SRR446031_2.fastq        V1A     V1 23-Mar-2012
## 6  ./data/SRR446032_1.fastq ./data/SRR446032_2.fastq        V1B     V1 23-Mar-2012
## 7  ./data/SRR446033_1.fastq ./data/SRR446033_2.fastq        M6A     M6 23-Mar-2012
## 8  ./data/SRR446034_1.fastq ./data/SRR446034_2.fastq        M6B     M6 23-Mar-2012
## 9  ./data/SRR446035_1.fastq ./data/SRR446035_2.fastq        A6A     A6 23-Mar-2012
## 10 ./data/SRR446036_1.fastq ./data/SRR446036_2.fastq        A6B     A6 23-Mar-2012
## 11 ./data/SRR446037_1.fastq ./data/SRR446037_2.fastq        V6A     V6 23-Mar-2012
## 12 ./data/SRR446038_1.fastq ./data/SRR446038_2.fastq        V6B     V6 23-Mar-2012
## 13 ./data/SRR446039_1.fastq ./data/SRR446039_2.fastq       M12A    M12 23-Mar-2012
## 14 ./data/SRR446040_1.fastq ./data/SRR446040_2.fastq       M12B    M12 23-Mar-2012
## 15 ./data/SRR446041_1.fastq ./data/SRR446041_2.fastq       A12A    A12 23-Mar-2012
## 16 ./data/SRR446042_1.fastq ./data/SRR446042_2.fastq       A12B    A12 23-Mar-2012
## 17 ./data/SRR446043_1.fastq ./data/SRR446043_2.fastq       V12A    V12 23-Mar-2012
## 18 ./data/SRR446044_1.fastq ./data/SRR446044_2.fastq       V12B    V12 23-Mar-2012
</code></p>

<h1 id="read-preprocessing">Read preprocessing</h1>

<h2 id="read-quality-filtering-and-trimming">Read quality filtering and trimming</h2>

<p>The following removes reads with low quality base calls (here Phred
scores below 20) from all FASTQ files.</p>

<p><code>r
args &lt;- systemArgs(sysma="param/trimPE.param", mytargets="targetsPE.txt")[1:4] # Note: subsetting!
filterFct &lt;- function(fq, cutoff=20, Nexceptions=0) {
    qcount &lt;- rowSums(as(quality(fq), "matrix") &lt;= cutoff)
    fq[qcount &lt;= Nexceptions] # Retains reads where Phred scores are &gt;= cutoff with N exceptions
}
preprocessReads(args=args, Fct="filterFct(fq, cutoff=20, Nexceptions=0)", batchsize=100000)
writeTargetsout(x=args, file="targets_PEtrim.txt", overwrite=TRUE)
</code></p>

<h2 id="fastq-quality-report">FASTQ quality report</h2>

<p>The following <code>seeFastq</code> and <code>seeFastqPlot</code> functions generate and plot a series of 
useful quality statistics for a set of FASTQ files including per cycle quality box
plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads
above quality cutoffs and mean quality distribution. The results are
written to a PDF file named <code>fastqReport.pdf</code>.</p>

<p><code>r
args &lt;- systemArgs(sysma="param/tophat.param", mytargets="targets.txt")
fqlist &lt;- seeFastq(fastq=infile1(args), batchsize=100000, klength=8)
pdf("./results/fastqReport.pdf", height=18, width=4*length(fqlist))
seeFastqPlot(fqlist)
dev.off()
</code></p>

<p><img src="fastqReport.png" alt="" /></p>
<div align="center">Figure 1: FASTQ quality report for 18 samples</div>
<p>&lt;/br&gt;</p>

<h1 id="alignments">Alignments</h1>

<h2 id="read-mapping-with-bwa-mem">Read mapping with <code>BWA-MEM</code></h2>

<p>The NGS reads of this project are aligned against the reference genome
sequence using the highly variant tolerant short read aligner <code>BWA-MEM</code>
[@Li2013-oy; @Li2009-oc]. The parameter settings of the aligner are
defined in the <code>bwa.param</code> file.</p>

<p><code>r
args &lt;- systemArgs(sysma="param/bwa.param", mytargets="targets.txt")
sysargs(args)[1] # Command-line parameters for first FASTQ file
</code></p>

<p>Runs the alignments sequentially (<em>e.g.</em> on a single machine)</p>

<p><code>r
moduleload(modules(args))
system("bwa index -a bwtsw ./data/tair10.fasta")
bampaths &lt;- runCommandline(args=args)
</code></p>

<p>Alternatively, the alignment jobs can be submitted to a compute cluster,
here using 72 CPU cores (18 <code>qsub</code> processes each with 4 CPU cores).</p>

<p><code>r
moduleload(modules(args))
system("bwa index -a bwtsw ./data/tair10.fasta")
resources &lt;- list(walltime="20:00:00", nodes=paste0("1:ppn=", cores(args)), memory="10gb")
reg &lt;- clusterRun(args, conffile=".BatchJobs.R", template="torque.tmpl", Njobs=18, runid="01", 
                  resourceList=resources)
waitForJobs(reg)
writeTargetsout(x=args, file="targets_bam.txt", overwrite=TRUE)
</code></p>

<p>Check whether all BAM files have been created</p>

<p><code>r
file.exists(outpaths(args))
</code></p>

<h2 id="read-mapping-with-gsnap">Read mapping with <code>gsnap</code></h2>

<p>An alternative variant tolerant aligner is <code>gsnap</code> from the <code>gmapR</code> package
[@Wu2010-iq]. The following code shows how to run this aligner on
multiple nodes of a computer cluster that uses Torque as scheduler.</p>

<p><code>r
library(gmapR); library(BiocParallel); library(BatchJobs)
args &lt;- systemArgs(sysma="param/gsnap.param", mytargets="targetsPE.txt")
gmapGenome &lt;- GmapGenome(systemPipeR::reference(args), directory="data", name="gmap_tair10chr", create=TRUE)
f &lt;- function(x) {
    library(gmapR); library(systemPipeR)
    args &lt;- systemArgs(sysma="param/gsnap.param", mytargets="targetsPE.txt")
    gmapGenome &lt;- GmapGenome(reference(args), directory="data", name="gmap_tair10chr", create=FALSE)
    p &lt;- GsnapParam(genome=gmapGenome, unique_only=TRUE, molecule="DNA", max_mismatches=3)
    o &lt;- gsnap(input_a=infile1(args)[x], input_b=infile2(args)[x], params=p, output=outfile1(args)[x])
}
funs &lt;- makeClusterFunctionsTorque("torque.tmpl")
param &lt;- BatchJobsParam(length(args), resources=list(walltime="20:00:00", nodes="1:ppn=1", memory="6gb"), cluster.functions=funs)
register(param)
d &lt;- bplapply(seq(along=args), f)
writeTargetsout(x=args, file="targets_gsnap_bam.txt", overwrite=TRUE)
</code></p>

<h2 id="read-and-alignment-stats">Read and alignment stats</h2>

<p>The following generates a summary table of the number of reads in each
sample and how many of them aligned to the reference.</p>

<p><code>r
read_statsDF &lt;- alignStats(args=args) 
write.table(read_statsDF, "results/alignStats.xls", row.names=FALSE, quote=FALSE, sep="\t")
</code></p>

<h2 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</h2>

<p>The <code>symLink2bam</code> function creates symbolic links to view the BAM alignment files in a
genome browser such as IGV. The corresponding URLs are written to a file
with a path specified under <code>urlfile</code>, here <code>IGVurl.txt</code>.</p>

<p><code>r
symLink2bam(sysargs=args, htmldir=c("~/.html/", "projects/gen242/"), 
            urlbase="http://biocluster.ucr.edu/~tgirke/", 
            urlfile="./results/IGVurl.txt")
</code></p>

<h1 id="variant-calling">Variant calling</h1>

<p>The following performs variant calling with <code>GATK</code>, <code>BCFtools</code> and <code>VariantTools</code> 
in parallel mode on a compute cluster [@McKenna2010-ql; @Li2011-ll]. If a cluster is not
available, the <code>runCommandline</code> function can be used to run the variant calling with <code>GATK</code> 
and <code>BCFtools</code> for each sample sequentially on a single machine, or <code>callVariants</code> in case 
of <code>VariantTools</code>. Typically, the user would choose here only one variant caller rather
than running several ones.</p>

<h2 id="variant-calling-with-gatk">Variant calling with <code>GATK</code></h2>

<p>The following creates in the inital step a new <code>targets</code> file
(<code>targets_bam.txt</code>). The first column of this file gives the paths to
the BAM files created in the alignment step. The new targets file and the
parameter file <code>gatk.param</code> are used to create a new <code>SYSargs</code>
instance for running GATK. Since GATK involves many processing steps, it is
executed by a bash script <code>gatk_run.sh</code> where the user can specify the
detailed run parameters. All three files are expected to be located in the
current working directory. Samples files for <code>gatk.param</code> and
<code>gatk_run.sh</code> are available in the <code>param</code> subdirectory
provided by <code>systemPipeRdata</code>.</p>

<p><code>r
moduleload("picard/1.130")
system("picard CreateSequenceDictionary R=./data/tair10.fasta O=./data/tair10.dict")
args &lt;- systemArgs(sysma="param/gatk.param", mytargets="targets_bam.txt")
resources &lt;- list(walltime="20:00:00", nodes=paste0("1:ppn=", 1), memory="10gb")
reg &lt;- clusterRun(args, conffile=".BatchJobs.R", template="torque.tmpl", Njobs=18, runid="01",
                  resourceList=resources)
waitForJobs(reg)
# unlink(outfile1(args), recursive = TRUE, force = TRUE)
writeTargetsout(x=args, file="targets_gatk.txt", overwrite=TRUE)
</code></p>

<h2 id="variant-calling-with-bcftools">Variant calling with <code>BCFtools</code></h2>

<p>The following runs the variant calling with <code>BCFtools</code>. This step requires
in the current working directory the parameter file <code>sambcf.param</code> and the bash script 
<code>sambcf_run.sh</code>.</p>

<p><code>r
args &lt;- systemArgs(sysma="param/sambcf.param", mytargets="targets_bam.txt")
resources &lt;- list(walltime="20:00:00", nodes=paste0("1:ppn=", 1), memory="10gb")
reg &lt;- clusterRun(args, conffile=".BatchJobs.R", template="torque.tmpl", Njobs=18, runid="01",
                  resourceList=resources)
waitForJobs(reg)
# unlink(outfile1(args), recursive = TRUE, force = TRUE)
writeTargetsout(x=args, file="targets_sambcf.txt", overwrite=TRUE)
</code></p>

<h2 id="variant-calling-with-varianttools">Variant calling with <code>VariantTools</code></h2>

<p><code>r
library(gmapR); library(BiocParallel); library(BatchJobs)
args &lt;- systemArgs(sysma="param/vartools.param", mytargets="targets_gsnap_bam.txt")
f &lt;- function(x) {
    library(VariantTools); library(gmapR); library(systemPipeR)
    args &lt;- systemArgs(sysma="param/vartools.param", mytargets="targets_gsnap_bam.txt")
    gmapGenome &lt;- GmapGenome(systemPipeR::reference(args), directory="data", name="gmap_tair10chr", create=FALSE)
    tally.param &lt;- TallyVariantsParam(gmapGenome, high_base_quality = 23L, indels = TRUE)
    bfl &lt;- BamFileList(infile1(args)[x], index=character())
    var &lt;- callVariants(bfl[[1]], tally.param)
    sampleNames(var) &lt;- names(bfl)
    writeVcf(asVCF(var), outfile1(args)[x], index = TRUE)
}
funs &lt;- makeClusterFunctionsTorque("torque.tmpl")
param &lt;- BatchJobsParam(length(args), resources=list(walltime="20:00:00", nodes="1:ppn=1", memory="6gb"), cluster.functions=funs)
register(param)
d &lt;- bplapply(seq(along=args), f)
writeTargetsout(x=args, file="targets_vartools.txt", overwrite=TRUE)
</code></p>

<h2 id="inspect-vcf-file">Inspect VCF file</h2>

<p>VCF files can be imported into R with the <code>readVcf</code> function. Both <code>VCF</code> and <code>VRanges</code> objects provide
convenient data structure for working with variant data (<em>e.g.</em> SNP quality filtering).</p>

<p><code>r
library(VariantAnnotation)
args &lt;- systemArgs(sysma="param/filter_gatk.param", mytargets="targets_gatk.txt")
vcf &lt;- readVcf(infile1(args)[1], "A. thaliana")
vcf
vr &lt;- as(vcf, "VRanges")
vr
</code></p>

<h1 id="filter-variants">Filter variants</h1>

<p>The function <code>filterVars</code> filters VCF files based on user definable
quality parameters. It sequentially imports each VCF file into R, applies the
filtering on an internally generated <code>VRanges</code> object and then writes
the results to a new subsetted VCF file. The filter parameters are passed on to
the corresponding argument as a character string. The function applies this
filter to the internally generated <code>VRanges</code> object using the standard
subsetting syntax for two dimensional objects such as: <code>vr[filter, ]</code>.
The parameter files (<code>filter_gatk.param</code>, <code>filter_sambcf.param</code> and 
<code>filter_vartools.param</code>), used in the filtering steps, define the paths to 
the input and output VCF files which are stored in new <code>SYSargs</code> instances.</p>

<h2 id="filter-variants-called-by-gatk">Filter variants called by <code>GATK</code></h2>

<p>The below example filters for variants that are supported by $\ge$<code>x</code>
reads and $\ge$80% of them support the called variants. In addition, all
variants need to pass $\ge$<code>x</code> of the soft filters recorded in the VCF
files generated by GATK. Since the toy data used for this workflow is
very small, the chosen settings are unreasonabley relaxed. A more
reasonable filter setting is given in the line below (here commented
out).</p>

<p><code>r
library(VariantAnnotation)
library(BBmisc) # Defines suppressAll()
args &lt;- systemArgs(sysma="param/filter_gatk.param", mytargets="targets_gatk.txt")[1:4]
filter &lt;- "totalDepth(vr) &gt;= 2 &amp; (altDepth(vr) / totalDepth(vr) &gt;= 0.8) &amp; rowSums(softFilterMatrix(vr))&gt;=1"
# filter &lt;- "totalDepth(vr) &gt;= 20 &amp; (altDepth(vr) / totalDepth(vr) &gt;= 0.8) &amp; rowSums(softFilterMatrix(vr))==6"
suppressAll(filterVars(args, filter, varcaller="gatk", organism="A. thaliana"))
writeTargetsout(x=args, file="targets_gatk_filtered.txt", overwrite=TRUE)
</code></p>

<h2 id="filter-variants-called-by-bcftools">Filter variants called by <code>BCFtools</code></h2>

<p>The following shows how to filter the VCF files generated by <code>BCFtools</code> using
similar parameter settings as in the previous filtering of the GATK
results.</p>

<p><code>r
args &lt;- systemArgs(sysma="param/filter_sambcf.param", mytargets="targets_sambcf.txt")[1:4]
filter &lt;- "rowSums(vr) &gt;= 2 &amp; (rowSums(vr[,3:4])/rowSums(vr[,1:4]) &gt;= 0.8)"
# filter &lt;- "rowSums(vr) &gt;= 20 &amp; (rowSums(vr[,3:4])/rowSums(vr[,1:4]) &gt;= 0.8)"
suppressAll(filterVars(args, filter, varcaller="bcftools", organism="A. thaliana"))
writeTargetsout(x=args, file="targets_sambcf_filtered.txt", overwrite=TRUE)
</code></p>

<h2 id="filter-variants-called-by-varianttools">Filter variants called by <code>VariantTools</code></h2>

<p>The following shows how to filter the VCF files generated by <code>VariantTools</code> using
similar parameter settings as in the previous filtering of the GATK
results.</p>

<p><code>r
library(VariantAnnotation)
library(BBmisc) # Defines suppressAll()
args &lt;- systemArgs(sysma="param/filter_vartools.param", mytargets="targets_vartools.txt")[1:4]
filter &lt;- "(values(vr)$n.read.pos.ref + values(vr)$n.read.pos) &gt;= 2 &amp; (values(vr)$n.read.pos / (values(vr)$n.read.pos.ref + values(vr)$n.read.pos) &gt;= 0.8)"
# filter &lt;- "(values(vr)$n.read.pos.ref + values(vr)$n.read.pos) &gt;= 20 &amp; (values(vr)$n.read.pos / (values(vr)$n.read.pos.ref + values(vr)$n.read.pos) &gt;= 0.8)"
filterVars(args, filter, varcaller="vartools", organism="A. thaliana")
writeTargetsout(x=args, file="targets_vartools_filtered.txt", overwrite=TRUE)
</code></p>

<p>Check filtering outcome for one sample</p>

<p><code>r
length(as(readVcf(infile1(args)[1], genome="Ath"), "VRanges")[,1])
length(as(readVcf(outpaths(args)[1], genome="Ath"), "VRanges")[,1])
</code></p>

<h1 id="annotate-filtered-variants">Annotate filtered variants</h1>

<p>The function <code>variantReport</code> generates a variant report using
utilities provided by the <code>VariantAnnotation</code> package. The report for
each sample is written to a tabular file containing genomic context annotations
(<em>e.g.</em> coding or non-coding SNPs, amino acid changes, IDs of affected
genes, etc.) along with confidence statistics for each variant. The parameter
file <code>annotate_vars.param</code> defines the paths to the input and output
files which are stored in a new <code>SYSargs</code> instance.</p>

<h2 id="basics-of-annotating-variants">Basics of annotating variants</h2>

<p>Variants overlapping with common annotation features can be identified with <code>locateVariants</code>.</p>

<p><code>r
library("GenomicFeatures")
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_gatk_filtered.txt")
txdb &lt;- loadDb("./data/tair10.sqlite")
vcf &lt;- readVcf(infile1(args)[1], "A. thaliana")
locateVariants(vcf, txdb, CodingVariants())
</code>
Synonymous/non-synonymous variants of coding sequences are computed by the predictCoding function for variants overlapping with coding regions.</p>

<p><code>r
fa &lt;- FaFile(systemPipeR::reference(args))
predictCoding(vcf, txdb, seqSource=fa)
</code></p>

<h2 id="annotate-filtered-variants-called-by-gatk">Annotate filtered variants called by <code>GATK</code></h2>

<p><code>r
library("GenomicFeatures")
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_gatk_filtered.txt")
txdb &lt;- loadDb("./data/tair10.sqlite")
fa &lt;- FaFile(systemPipeR::reference(args))
suppressAll(variantReport(args=args, txdb=txdb, fa=fa, organism="A. thaliana"))
</code></p>

<h2 id="annotate-filtered-variants-called-by-bcftools">Annotate filtered variants called by <code>BCFtools</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
txdb &lt;- loadDb("./data/tair10.sqlite")
fa &lt;- FaFile(systemPipeR::reference(args))
suppressAll(variantReport(args=args, txdb=txdb, fa=fa, organism="A. thaliana"))
</code></p>

<h2 id="annotate-filtered-variants-called-by-varianttools">Annotate filtered variants called by <code>VariantTools</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_vartools_filtered.txt")
txdb &lt;- loadDb("./data/tair10.sqlite")
fa &lt;- FaFile(systemPipeR::reference(args))
suppressAll(variantReport(args=args, txdb=txdb, fa=fa, organism="A. thaliana"))
</code></p>

<p>View annotation result for single sample</p>

<p><code>r
read.delim(outpaths(args)[1])[38:40,]
</code></p>

<h1 id="combine-annotation-results-among-samples">Combine annotation results among samples</h1>

<p>To simplify comparisons among samples, the \Rfunction{combineVarReports}
function combines all variant annotation reports referenced in a
<code>SYSargs</code> instance (here <code>args</code>). At the same time the function
allows to consider only certain feature types of interest. For instance, the
below setting <code>filtercol=c(Consequence="nonsynonymous")</code> will include
only nonsysynonymous variances listed in the <code>Consequence</code> column of
the annotation reports. To omit filtering, one can use the setting
<code>filtercol="All"</code>.</p>

<h2 id="combine-results-from-gatk">Combine results from <code>GATK</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_gatk_filtered.txt")
combineDF &lt;- combineVarReports(args, filtercol=c(Consequence="nonsynonymous"))
write.table(combineDF, "./results/combineDF_nonsyn_gatk.xls", quote=FALSE, row.names=FALSE, sep="\t")
</code></p>

<h2 id="combine-results-from-bcftools">Combine results from <code>BCFtools</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
combineDF &lt;- combineVarReports(args, filtercol=c(Consequence="nonsynonymous"))
write.table(combineDF, "./results/combineDF_nonsyn_sambcf.xls", quote=FALSE, row.names=FALSE, sep="\t")
</code></p>

<h2 id="combine-results-from-varianttools">Combine results from <code>VariantTools</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_vartools_filtered.txt")
combineDF &lt;- combineVarReports(args, filtercol=c(Consequence="nonsynonymous"))
write.table(combineDF, "./results/combineDF_nonsyn_vartools.xls", quote=FALSE, row.names=FALSE, sep="\t")
combineDF[2:4,] 
</code></p>

<h1 id="summary-statistics-of-variants">Summary statistics of variants</h1>

<p>The <code>varSummary</code> function counts the number of variants for each feature type
included in the anntation reports.</p>

<h2 id="summary-for-gatk">Summary for <code>GATK</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_gatk_filtered.txt")
varSummary(args)
write.table(varSummary(args), "./results/variantStats_gatk.xls", quote=FALSE, col.names = NA, sep="\t")
</code></p>

<h2 id="summary-for-bcftools">Summary for <code>BCFtools</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
varSummary(args)
write.table(varSummary(args), "./results/variantStats_sambcf.xls", quote=FALSE, col.names = NA, sep="\t")
</code></p>

<h2 id="summary-for-varianttools">Summary for <code>VariantTools</code></h2>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_vartools_filtered.txt")
varSummary(args)
write.table(varSummary(args), "./results/variantStats_vartools.xls", quote=FALSE, col.names = NA, sep="\t")
</code></p>

<h1 id="venn-diagram-of-variants">Venn diagram of variants</h1>

<p>The venn diagram utilities defined by the <code>systemPipeR</code> package can be used to
identify common and unique variants reported for different samples
and/or variant callers. The below generates a 4-way venn diagram
comparing four sampes for each of the two variant callers.</p>

<p><code>r
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_gatk_filtered.txt")
varlist &lt;- sapply(names(outpaths(args))[1:4], function(x) as.character(read.delim(outpaths(args)[x])$VARID))
vennset_gatk &lt;- overLapper(varlist, type="vennsets")
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_sambcf_filtered.txt")
varlist &lt;- sapply(names(outpaths(args))[1:4], function(x) as.character(read.delim(outpaths(args)[x])$VARID))
vennset_bcf &lt;- overLapper(varlist, type="vennsets")
args &lt;- systemArgs(sysma="param/annotate_vars.param", mytargets="targets_vartools_filtered.txt")
varlist &lt;- sapply(names(outpaths(args))[1:4], function(x) as.character(read.delim(outpaths(args)[x])$VARID))
vennset_vartools &lt;- overLapper(varlist, type="vennsets")
pdf("./results/vennplot_var.pdf")
vennPlot(list(vennset_gatk, vennset_bcf, vennset_vartools), mymain="", mysub="GATK: red; BCFtools: blue; VariantTools: green", colmode=2, ccol=c("red", "blue", "green"))
dev.off()
</code></p>

<p><img src="vennplot_var.png" alt="" /></p>
<div align="center">Figure 2: Venn Diagram for 4 samples from GATK and BCFtools</div>
<p>&lt;/br&gt;</p>

<h1 id="plot-variants-programmatically">Plot variants programmatically</h1>

<p>The following plots a selected variant with <code>ggbio</code>.</p>

<p><code>r
library(ggbio)
mychr &lt;- "ChrC"; mystart &lt;- 11000; myend &lt;- 13000
args &lt;- systemArgs(sysma="param/bwa.param", mytargets="targets.txt")
ga &lt;- readGAlignments(outpaths(args)[1], use.names=TRUE, param=ScanBamParam(which=GRanges(mychr, IRanges(mystart, myend))))
p1 &lt;- autoplot(ga, geom = "rect")
p2 &lt;- autoplot(ga, geom = "line", stat = "coverage")
p3 &lt;- autoplot(vcf[seqnames(vcf)==mychr], type = "fixed") + xlim(mystart, myend) + theme(legend.position = "none", axis.text.y = element_blank(), axis.ticks.y=element_blank())
p4 &lt;- autoplot(txdb, which=GRanges(mychr, IRanges(mystart, myend)), names.expr = "gene_id")
png("./results/plot_variant.png")
tracks(Reads=p1, Coverage=p2, Variant=p3, Transcripts=p4, heights = c(0.3, 0.2, 0.1, 0.35)) + ylab("")
dev.off()
</code></p>

<p><img src="plot_variant.png" alt="" /></p>
<div align="center">Figure 3: Plot variants with programmatically.</div>
<p>&lt;/br&gt;</p>

<h1 id="version-information">Version Information</h1>

<p><code>r
sessionInfo()
</code></p>

<p><code>
## R version 3.2.5 (2016-04-14)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 14.04.4 LTS
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  utils     datasets  grDevices methods   base     
## 
## other attached packages:
##  [1] ape_3.4                    ggplot2_2.0.0              systemPipeR_1.4.8         
##  [4] RSQLite_1.0.0              DBI_0.3.1                  ShortRead_1.28.0          
##  [7] GenomicAlignments_1.6.1    SummarizedExperiment_1.0.1 Biobase_2.30.0            
## [10] BiocParallel_1.4.0         Rsamtools_1.22.0           Biostrings_2.38.2         
## [13] XVector_0.10.0             GenomicRanges_1.22.1       GenomeInfoDb_1.6.1        
## [16] IRanges_2.4.4              S4Vectors_0.8.3            BiocGenerics_0.16.1       
## [19] BiocStyle_1.8.0           
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.3            lattice_0.20-33        GO.db_3.2.2            digest_0.6.9          
##  [5] plyr_1.8.3             futile.options_1.0.0   BatchJobs_1.6          evaluate_0.8          
##  [9] zlibbioc_1.16.0        GenomicFeatures_1.22.6 annotate_1.48.0        Matrix_1.2-3          
## [13] checkmate_1.6.3        rmarkdown_0.9.2        GOstats_2.36.0         splines_3.2.5         
## [17] stringr_1.0.0          pheatmap_1.0.7         RCurl_1.95-4.7         biomaRt_2.26.1        
## [21] munsell_0.4.2          sendmailR_1.2-1        rtracklayer_1.30.1     base64enc_0.1-3       
## [25] BBmisc_1.9             htmltools_0.3          fail_1.3               edgeR_3.12.0          
## [29] codetools_0.2-14       XML_3.98-1.3           AnnotationForge_1.12.0 bitops_1.0-6          
## [33] grid_3.2.5             RBGL_1.46.0            nlme_3.1-122           xtable_1.8-0          
## [37] GSEABase_1.32.0        gtable_0.1.2           magrittr_1.5           formatR_1.2.1         
## [41] scales_0.3.0           graph_1.48.0           stringi_1.0-1          hwriter_1.3.2         
## [45] genefilter_1.52.0      limma_3.26.3           latticeExtra_0.6-26    futile.logger_1.4.1   
## [49] brew_1.0-6             rjson_0.2.15           lambda.r_1.1.7         RColorBrewer_1.1-2    
## [53] tools_3.2.5            Category_2.36.0        survival_2.38-3        yaml_2.1.13           
## [57] AnnotationDbi_1.32.0   colorspace_1.2-6       knitr_1.12
</code></p>

<h1 id="funding">Funding</h1>

<p>This project was supported by funds from the National Institutes of
Health (NIH) and the National Science Foundation (NSF).</p>

<h1 id="references">References</h1>



<div class="tags">
    
</div>



</div>



<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2016 UC Riverside. All rights reserved. <br />
 Site last generated: May 24, 2016 <br />
<p><img src="../common_images/company_logo.png/></p>
                </div>
            </div>
</footer>

            </div><!-- /.row -->

    </div>    <!-- /.container -->

</body>


</html>

